<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width,minimum-scale=1">

  <title>Experience from LINCS</title>
  <meta name="description" content="# Experience from LINCSA cookbook recipe documenting the easy extract & transform (ETL) process of fitting the LINCS data into the C2M2 model.**Authors**: [D...">

  <link rel="canonical" href="https://nih-cfde.github.io/the-fair-cookbook/content/recipes/15/4/lincs.html">
  <link rel="alternate" type="application/rss+xml" title="NIH-CFDE FAIR COOKBOOK" href="https://nih-cfde.github.io/the-fair-cookbook/feed.xml">

  <meta property="og:url"         content="https://nih-cfde.github.io/the-fair-cookbook/content/recipes/15/4/lincs.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="Experience from LINCS" />
<meta property="og:description" content="# Experience from LINCSA cookbook recipe documenting the easy extract & transform (ETL) process of fitting the LINCS data into the C2M2 model.**Authors**: [D..." />
<meta property="og:image"       content="https://nih-cfde.github.io/the-fair-cookbook/images/logo/CFDE-logo-contrast.png" />

<meta name="twitter:card" content="summary">


  <script type="application/ld+json">
  {
  "@context": "http://schema.org",
  "@type": "NewsArticle",
  "mainEntityOfPage": "https://nih-cfde.github.io/the-fair-cookbook/content/recipes/15/4/lincs.html",
  "headline": "Experience from LINCS",
  "datePublished": "2020-07-09T20:59:51+00:00",
  "dateModified": "2020-07-09T20:59:51+00:00",
  "description": "# Experience from LINCSA cookbook recipe documenting the easy extract & transform (ETL) process of fitting the LINCS data into the C2M2 model.**Authors**: [D...",
  "author": {
    "@type": "Person",
    "name": "CFDE"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Data 100 at UC Berkeley",
    "logo": {
      "@type": "ImageObject",
      "url": "https://nih-cfde.github.io/the-fair-cookbook",
      "width": 60,
      "height": 60
    }
  },
  "image": {
    "@type": "ImageObject",
    "url": "https://nih-cfde.github.io/the-fair-cookbook",
    "height": 60,
    "width": 60
  }
}

  </script>
  <link rel="stylesheet" href="/the-fair-cookbook/assets/css/styles.css">

  <!-- <link rel="manifest" href="/manifest.json"> -->
  <!-- <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#efae0a"> -->
  <meta name="msapplication-TileColor" content="#da532c">
  <meta name="msapplication-TileImage" content="/mstile-144x144.png">
  <meta name="theme-color" content="#233947">

  <!-- Favicon -->
  <link rel="shortcut icon" type="image/x-icon" href="/the-fair-cookbook/images/logo/CFDE-logo.ico">

  <!-- MathJax Config -->
  <!-- Allow inline math using $ and automatically break long math lines -->
<!-- (mostly) copied from nbconvert configuration -->
<!-- https://github.com/jupyter/nbconvert/blob/master/nbconvert/templates/html/mathjax.tpl -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true },
    },
    
});
</script>
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS_HTML' async></script>


  <!-- DOM updating function -->
  <script src="/the-fair-cookbook/assets/js/page/dom-update.js"></script>

  <!-- Selectors for elements on the page -->
  <script src="/the-fair-cookbook/assets/js/page/documentSelectors.js"></script>

  <!-- Define some javascript variables that will be useful in other javascript -->
  <script>
    const site_basename = '/the-fair-cookbook';
  </script>

  <!-- Add AnchorJS to let headers be linked -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/4.2.0/anchor.min.js" async></script>
  <script src="/the-fair-cookbook/assets/js/page/anchors.js" async></script>

  <!-- Include Turbolinks to make page loads fast -->
  <!-- https://github.com/turbolinks/turbolinks -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/turbolinks/5.2.0/turbolinks.js" async></script>
  <meta name="turbolinks-cache-control" content="no-cache">

  <!-- Load nbinteract for widgets -->
  

  <!-- Load Thebelab for interactive widgets -->
  <!-- Include Thebelab for interactive code if it's enabled -->


<!-- Display Thebelab button in each code cell -->
<script>
/**
 * Set up thebelab button for code blocks
 */

const thebelabCellButton = id =>
  `<a id="thebelab-cell-button-${id}" class="btn thebebtn o-tooltip--left" data-tooltip="Interactive Mode">
    <img src="/the-fair-cookbook/assets/images/edit-button.svg" alt="Start thebelab interactive mode">
  </a>`


const addThebelabButtonToCodeCells =  () => {

  const codeCells = document.querySelectorAll('div.input_area > div.highlight:not(.output) pre')
  codeCells.forEach((codeCell, index) => {
    const id = codeCellId(index)
    codeCell.setAttribute('id', id)
    if (document.getElementById("thebelab-cell-button-" + id) == null) {
      codeCell.insertAdjacentHTML('afterend', thebelabCellButton(id));
    }
  })
}

initFunction(addThebelabButtonToCodeCells);
</script>


<script src="https://unpkg.com/thebelab@latest/lib/index.js" async></script>
<script>
    /**
     * Add attributes to Thebelab blocks
     */

    const initThebelab = () => {
        const addThebelabToCodeCells = () => {
            console.log("Adding thebelab to code cells...");
            // If Thebelab hasn't loaded, wait a bit and try again. This
            // happens because we load ClipboardJS asynchronously.
            if (window.thebelab === undefined) {
                setTimeout(addThebelabToCodeCells, 250)
            return
            }

            // If we already detect a Thebelab cell, don't re-run
            if (document.querySelectorAll('div.thebelab-cell').length > 0) {
                return;
            }

            // Find all code cells, replace with Thebelab interactive code cells
            const codeCells = document.querySelectorAll('.input_area pre')
            codeCells.forEach((codeCell, index) => {
                const id = codeCellId(index)

                // Clean up the language to make it work w/ CodeMirror and add it to the cell
                dataLanguage = ""
                dataLanguage = detectLanguage(dataLanguage);
                codeCell.setAttribute('data-language', dataLanguage)
                codeCell.setAttribute('data-executable', 'true')

                // If the code cell is hidden, show it
                var inputCheckbox = document.querySelector(`input#hidebtn${codeCell.id}`);
                if (inputCheckbox !== null) {
                    setCodeCellVisibility(inputCheckbox, 'visible');
                }
            });

            // Remove the event listener from the page so keyboard press doesn't
            // Change page
            document.removeEventListener('keydown', initPageNav)
            keyboardListener = false;

            // Init thebelab
            thebelab.bootstrap();

            // Remove copy buttons since they won't work anymore
            const copyAndThebeButtons = document.querySelectorAll('.copybtn, .thebebtn')
            copyAndThebeButtons.forEach((button, index) => {
                button.remove();
            });

            // Remove outputs since they'll be stale
            const outputs = document.querySelectorAll('.output *, .output')
            outputs.forEach((output, index) => {
                output.remove();
            });

            // Find any cells with an initialization tag and ask ThebeLab to run them when ready
            var thebeInitCells = document.querySelectorAll('div.tag_thebelab-init');
            thebeInitCells.forEach((cell) => {
                console.log("Initializing ThebeLab with cell: " + cell.id);
                cell.querySelector('.thebelab-run-button').click();
            });
        }

        // Add event listener for the function to modify code cells
        const thebelabButtons = document.querySelectorAll('[id^=thebelab], [id$=thebelab]')
        thebelabButtons.forEach((thebelabButton,index) => {
            if (thebelabButton === null) {
                setTimeout(initThebelab, 250)
                return
            };
            thebelabButton.addEventListener('click', addThebelabToCodeCells);
        });
    }

    // Initialize Thebelab
    initFunction(initThebelab);

// Helper function to munge the language name
var detectLanguage = (language) => {
    if (language.indexOf('python') > -1) {
        language = "python";
    }
    return language;
}
</script>



  <!-- Load the auto-generating TOC (non-async otherwise the TOC won't load w/ turbolinks) -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.8.1/tocbot.min.js" async></script>
  <script src="/the-fair-cookbook/assets/js/page/tocbot.js"></script>



<!-- using mermaid for rendering Diagram and flow chart -->
<!-- <script src="https://cdnjs.cloudflare.com/ajax/libs/mermaid/8.4.8/mermaid.min.js" async></script> -->
<script src="/the-fair-cookbook/assets/js/mermaid.min.js"></script>
<script>mermaid.initialize({startOnLoad:true});</script>
<!--  <script src="/the-fair-cookbook/assets/js/mermaid.js"></script>-->



  <!-- Google analytics -->
  
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id="></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '');
</script>



  <!-- Clipboard copy button -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" async></script>

  <!-- Load custom website scripts -->
  <script src="/the-fair-cookbook/assets/js/scripts.js" async></script>

  <!-- Load custom user CSS and JS  -->
  <script src="/the-fair-cookbook/assets/custom/custom.js" async></script>
  <link rel="stylesheet" href="/the-fair-cookbook/assets/custom/custom.css">

  <!-- Update interact links w/ REST param, is defined in includes so we can use templates -->
  
<script>
/**
  * To auto-embed hub URLs in interact links if given in a RESTful fashion
 */

function getJsonFromUrl(url) {
  var query = url.split('?');
  if (query.length < 2) {
    // No queries so just return false
    return false;
  }
  query = query[1];
  // Collect REST params into a dictionary
  var result = {};
  query.split("&").forEach(function(part) {
    var item = part.split("=");
    result[item[0]] = decodeURIComponent(item[1]);
  });
  return result;
}
    
function dict2param(dict) {
    params = Object.keys(dict).map(function(k) {
        return encodeURIComponent(k) + '=' + encodeURIComponent(dict[k])
    });
    return params.join('&')
}

// Parse a Binder URL, converting it to the string needed for JupyterHub
function binder2Jupyterhub(url) {
  newUrl = {};
  parts = url.split('v2/gh/')[1];
  // Grab the base repo information
  repoinfo = parts.split('?')[0];
  var [org, repo, ref] = repoinfo.split('/');
  newUrl['repo'] = ['https://github.com', org, repo].join('/');
  newUrl['branch'] = ref
  // Grab extra parameters passed
  params = getJsonFromUrl(url);
  if (params['filepath'] !== undefined) {
    newUrl['subPath'] = params['filepath']
  }
  return dict2param(newUrl);
}

// Filter out potentially unsafe characters to prevent xss
function safeUrl(url)
{
   return String(encodeURIComponent(url))
            .replace(/&/g, '&amp;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&#39;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;');
}

function addParamToInternalLinks(hub) {
  var links = document.querySelectorAll("a").forEach(function(link) {
    var href = link.href;
    // If the link is an internal link...
    if (href.search("https://nih-cfde.github.io") !== -1 || href.startsWith('/') || href.search("127.0.0.1:") !== -1) {
      // Assume we're an internal link, add the hub param to it
      var params = getJsonFromUrl(href);
      if (params !== false) {
        // We have REST params, so append a new one
        params['jupyterhub'] = hub;
      } else {
        // Create the REST params
        params = {'jupyterhub': hub};
      }
      // Update the link
      var newHref = href.split('?')[0] + '?' + dict2param(params);
      link.setAttribute('href', decodeURIComponent(newHref));
    }
  });
  return false;
}


// Update interact links
function updateInteractLink() {
    // hack to make this work since it expects a ? in the URL
    rest = getJsonFromUrl("?" + location.search.substr(1));
    jupyterHubUrl = rest['jupyterhub'];
    var hubType = null;
    var hubUrl = null;
    if (jupyterHubUrl !== undefined) {
      hubType = 'jupyterhub';
      hubUrl = jupyterHubUrl;
    }

    if (hubType !== null) {
      // Sanitize the hubUrl
      hubUrl = safeUrl(hubUrl);

      // Add HTTP text if omitted
      if (hubUrl.indexOf('http') < 0) {hubUrl = 'http://' + hubUrl;}
      var interactButtons = document.querySelectorAll("button.interact-button")
      var lastButton = interactButtons[interactButtons.length-1];
      var link = lastButton.parentElement;

      // If we've already run this, skip the link updating
      if (link.nextElementSibling !== null) {
        return;
      }

      // Update the link and add context div
      var href = link.getAttribute('href');
      if (lastButton.id === 'interact-button-binder') {
        // If binder links exist, we need to re-work them for jupyterhub
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // If localhost, assume we're working from a local Jupyter server and remove `/hub`
          first = [hubUrl, 'git-sync'].join('/')
        } else {
          first = [hubUrl, 'hub', 'user-redirect', 'git-sync'].join('/')
        }
        href = first + '?' + binder2Jupyterhub(href);
      } else {
        // If interact button isn't binderhub, assume it's jupyterhub
        // If JupyterHub links, we only need to replace the hub url
        href = href.replace("", hubUrl);
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // Assume we're working from a local Jupyter server and remove `/hub`
          href = href.replace("/hub/user-redirect", "");
        }
      }
      link.setAttribute('href', decodeURIComponent(href));

      // Add text after interact link saying where we're launching
      hubUrlNoHttp = decodeURIComponent(hubUrl).replace('http://', '').replace('https://', '');
      link.insertAdjacentHTML('afterend', '<div class="interact-context">on ' + hubUrlNoHttp + '</div>');

      // Update internal links so we retain the hub url
      addParamToInternalLinks(hubUrl);
    }
}

runWhenDOMLoaded(updateInteractLink)
document.addEventListener('turbolinks:load', updateInteractLink)
</script>


  <!-- Lunr search code - will only be executed on the /search page -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.6/lunr.min.js" async></script>
  <script>var initQuery = function() {
  // See if we have a search box
  var searchInput = document.querySelector('input#lunr_search');
  if (searchInput === null) {
    return;
  }

  // Function to parse our lunr cache
  var idx = lunr(function () {
    this.field('title')
    this.field('excerpt')
    this.field('categories')
    this.field('tags')
    this.ref('id')

    this.pipeline.remove(lunr.trimmer)

    for (var item in store) {
      this.add({
        title: store[item].title,
        excerpt: store[item].excerpt,
        categories: store[item].categories,
        tags: store[item].tags,
        id: item
      })
    }
  });

  // Run search upon keyup
  searchInput.addEventListener('keyup', function () {
    var resultdiv = document.querySelector('#results');
    var query = document.querySelector("input#lunr_search").value.toLowerCase();
    var result =
      idx.query(function (q) {
        query.split(lunr.tokenizer.separator).forEach(function (term) {
          q.term(term, { boost: 100 })
          if(query.lastIndexOf(" ") != query.length-1){
            q.term(term, {  usePipeline: false, wildcard: lunr.Query.wildcard.TRAILING, boost: 10 })
          }
          if (term != ""){
            q.term(term, {  usePipeline: false, editDistance: 1, boost: 1 })
          }
        })
      });

      // Empty the results div
      while (resultdiv.firstChild) {
        resultdiv.removeChild(resultdiv.firstChild);
      }

    resultdiv.insertAdjacentHTML('afterbegin', '<p class="results__found">'+result.length+' Result(s) found</p>');
    for (var item in result) {
      var ref = result[item].ref;
      if(store[ref].teaser){
        var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<div class="archive__item-teaser">'+
                '<img src="'+store[ref].teaser+'" alt="">'+
              '</div>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      else{
    	  var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      resultdiv.insertAdjacentHTML('beforeend', searchitem);
    }
  });
};

initFunction(initQuery);
</script>

  <!-- Load JS that depends on site variables -->
  <script src="/the-fair-cookbook/assets/js/page/copy-button.js" async></script>

  <!-- Hide cell code -->
  <script src="/the-fair-cookbook/assets/js/page/hide-cell.js" async></script>

  <!-- Printing the screen -->
  <!-- Include nbinteract for interactive widgets -->
<script src="https://printjs-4de6.kxcdn.com/print.min.js" async></script>
<script>
printContent = () => {
    // MathJax displays a second version of any math for assistive devices etc.
    // This prevents double-rendering in the PDF output.
    var ignoreAssistList = [];
    assistives = document.querySelectorAll('.MathJax_Display span.MJX_Assistive_MathML').forEach((element, index) => {
        var thisId = 'MathJax-assistive-' + index.toString();
        element.setAttribute('id', thisId);
        ignoreAssistList.push(thisId)
    });

    // Print the actual content object
    printJS({
        printable: 'textbook_content',
        type: 'html',
        css: "/the-fair-cookbook/assets/css/styles.css",
        style: "#textbook_content {padding-top: 40px};",
        scanStyles: false,
        targetStyles: ["*"],
        ignoreElements: ignoreAssistList,
        documentTitle: "Made with Jupyter Book"
    })
};

initPrint = () => {
    document.querySelector('#interact-button-print').addEventListener('click', printContent)
}

initFunction(initPrint)
</script>

</head>

  <body>
    <!-- Include the ThebeLab config so it gets reloaded on each page -->
    <script type="text/x-thebe-config">{
    requestKernel: true,
    binderOptions: {
    repo: "nih-cfde/the-fair-cookbook",
    ref: "gh-pages",
    },
    codeMirrorConfig: {
    theme: "abcdef",
    mode: "python"
    },
    kernelOptions: {
    kernelName: "python3",
    path: ""
    }
}
</script>

    <!-- .js-show-sidebar shows sidebar by default -->
    <div id="js-textbook" class="c-textbook js-show-sidebar">
      



<nav id="js-sidebar" class="c-textbook__sidebar">
  <a href="https://nih-cfde.github.io/the-fair-cookbook/intro.html"><img src="/the-fair-cookbook/images/logo/CFDE-logo-contrast.png" class="textbook_logo" id="sidebar-logo" alt="textbook logo" data-turbolinks-permanent/></a>
  <h2 class="c-sidebar__title">NIH-CFDE FAIR COOKBOOK</h2>
  <ul class="c-sidebar__chapters">
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/intro">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/intro.html"
        >
          
          Introduction
        </a>
      </li>

      
      

      

      
      

      
        

        

        <ul class="c-sidebar__sections u-hidden-visually">
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/00/fair-principles">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/00/fair-principles.html"
              >
                
                What is FAIR?
              </a>
            </li>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/05/cfde">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/05/cfde.html"
              >
                
                Engaging with NIH-CFDE
              </a>
            </li>
            
            
          
        </ul>
      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/recipes/recipes-overview">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/recipes/recipes-overview.html"
        >
          
          Recipes
        </a>
      </li>

      
      

      

      
      

      
        

        

        <ul class="c-sidebar__sections u-hidden-visually">
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/06/cataloguing">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/06/cataloguing.html"
              >
                
                Discoverability
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/12/c2m2">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/12/c2m2.html"
                  >
                    
                    Introducing the Cross-Cut Metadata Model (C2M2)
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/12/C2M2-L1-description">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/12/C2M2-L1-description.html"
                  >
                    
                    Description of the Level-1 C2M2
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/06/cfde-namespaces">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/06/cfde-namespaces.html"
                  >
                    
                    CFDE Resource Naming
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/07/seo">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/07/seo.html"
                  >
                    
                    ETL to C2M2/DATS - the KidsFirst example
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/15/5/kf">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/15/5/kf.html"
                  >
                    
                    ETL to C2M2- the KidsFirst example
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/15/4/lincs">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/15/4/lincs.html"
                  >
                    
                    ETL to C2M2- the LINCS example
                  </a>
                </li>
              
              </ul>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/08/identifiers">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/08/identifiers.html"
              >
                
                Identification
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/08/1/pids">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/08/1/pids.html"
                  >
                    
                    Minting persistent and resolvable identifiers
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/08/2/minids">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/08/2/minids.html"
                  >
                    
                    Using MINIDs
                  </a>
                </li>
              
              </ul>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/09/finding">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/09/finding.html"
              >
                
                Semantics
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/10/ontologies">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/10/ontologies.html"
                  >
                    
                    Understanding terminologies
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/11/onto-services">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/11/onto-services.html"
                  >
                    
                    Ontology services
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/14/cfde-terminologies">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/14/cfde-terminologies.html"
                  >
                    
                    NIH CFDE selected terminologies
                  </a>
                </li>
              
              </ul>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/fair-assessment">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/fair-assessment.html"
              >
                
                FAIR compliance
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/04/fairshake">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/04/fairshake.html"
                  >
                    
                    Evaluating FAIRness with FAIRshake
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/16/fair-api">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/16/fair-api.html"
                  >
                    
                    Developing FAIR API
                  </a>
                </li>
              
              </ul>
            
            
          
        </ul>
      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/CFDE_Glossary">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/CFDE_Glossary.html"
        >
          
          Glossary
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/contributing">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/contributing.html"
        >
          
          Contributing
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/governance">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/governance.html"
        >
          
          Governance
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/license">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/license.html"
        >
          
          License
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/disclaimer">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/disclaimer.html"
        >
          
          Disclaimer
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/code_of_conduct">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/code_of_conduct.html"
        >
          
          Code of Conduct
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="https://github.com/nih-cfde/the-fair-cookbook">
        <a class="c-sidebar__entry"
          href="https://github.com/nih-cfde/the-fair-cookbook"
        >
          
          GitHub repository
        </a>
      </li>

      
      

      

      
      

      

      
    
  </ul>
  <p class="sidebar_footer">Powered by <a href="https://jupyterbook.org">Jupyter Book</a></p>
</nav>

      
      <div class="c-topbar" id="top-navbar">
  <!-- We show the sidebar by default so we use .is-active -->
  <div class="c-topbar__buttons">
    <button
      id="js-sidebar-toggle"
      class="hamburger hamburger--arrowalt is-active"
    >
      <span class="hamburger-box">
        <span class="hamburger-inner"></span>
      </span>
    </button>
    <div class="buttons">
<div class="download-buttons-dropdown">
    <button id="dropdown-button-trigger" class="interact-button"><img src="/the-fair-cookbook/assets/images/download-solid.svg" alt="Download" /></button>
    <div class="download-buttons">
        
        <a id="interact-button-print"><button id="interact-button-download" class="interact-button">.pdf</button></a>
    </div>
</div>


</div>

  </div>
  <!-- Empty sidebar placeholder that we'll auto-fill with javascript -->
  <aside class="sidebar__right">
    <header><h4 class="nav__title"><img src="/the-fair-cookbook/assets/images/list-solid.svg" alt="Search" />   On this page</h4></header>
    <nav class="onthispage">
    </nav>
  </aside>
  <a href="/the-fair-cookbook/search.html" class="topbar-right-button" id="search-button">
    <img src="/the-fair-cookbook/assets/images/search-solid.svg" alt="Search" />
  </a>
</div>

      <main class="c-textbook__page" tabindex="-1">
            <div class="c-textbook__content" id="textbook_content">
              <h1 id="experience-from-lincs">Experience from LINCS</h1>

<p>A cookbook recipe documenting the easy extract &amp; transform (ETL) process of fitting the LINCS data into the C2M2 model.</p>

<p><strong>Authors</strong>: <a href="https://orcid.org/0000-0003-3471-7416">Daniel J. B. Clarke</a></p>

<p><strong>Maintainers</strong>: <a href="https://orcid.org/0000-0003-3471-7416">Daniel J. B. Clarke</a></p>

<p><strong>Version</strong>: 1.0</p>

<p><strong>License</strong>: <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0 Universal (CC0 1.0) Public Domain Dedication</a></p>

<h2 id="objectives">Objectives</h2>
<ul>
  <li>Demonstrate the ETL process for preparing data for the C2M2 Level 1</li>
  <li>Introduce tooling to make this process easier</li>
  <li>Discuss necessary tradeoffs made to ensure maximal harmonization</li>
</ul>

<h2 id="introduction">Introduction</h2>

<p>The <a href="http://lincsportal.ccs.miami.edu/dcic-portal/">LINCS Data Portal</a> provides various ways to browse the existing LINCS data and highlight which small molecules, cell lines, genes, proteins, and antibodies are featured in a given dataset.</p>

<p>While LINCS revolves around <em>Datasets</em>, encompassing in some cases several files for each data level, the <a href="https://www.nih-cfde.org/product/cfde-c2m2/">C2M2</a> models information revolving around files. Fortunately, the <a href="https://www.nih-cfde.org/product/level-1-asset-manifest-specification/">C2M2 Level 1</a> provides collections, which enable arbitrary meaningful groupings of files, which is how the LINCS datasets can fit.</p>

<h2 id="step-by-step-process">Step by Step Process</h2>

<h3 id="step-1-setting-up-your-environment">Step 1: Setting up your environment</h3>

<p>The C2M2 Level 1 is described using a <a href="http://frictionlessdata.io/data-package/">Frictionless datapackage specification</a>, which permits various additional tooling to be devised around it. In the case that your data is already accessible in a large table, it may, in some cases, be simpler to construct mysql views from your data that are oriented in a C2M2 compliant manner and can be validated with datapackages. Here however, we will demonstrate producing a C2M2-compliant datapackage directly from the LINCS REST API.</p>

<p>A python helper class was devised to simplify codification of C2M2-compliant datapackages using <a href="https://docs.python.org/3/library/dataclasses.html">python3.7’s dataclasses</a> which provide convenient syntax highlighting and runtime assertions to catch errors early and easily introspect the C2M2 model with python docstrings. This package is available <a href="https://github.com/nih-cfde/FAIR/tree/master/Demos/FrictionlessDataclass">here</a> and is easily updated to future C2M2 schemas. This can be installed directly from GitHub with:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pip <span class="nb">install </span>git+git://github.com/nih-cfde/FAIR.git#egg<span class="o">=</span>c2m2_frictionless&amp;subdirectory<span class="o">=</span>Demos/FrictionlessDataclass
</code></pre></div></div>

<p>This is also the time to install any relevant packages for interacting with your DCC’s API, in our case we will use urllib to access the REST API.</p>

<h3 id="step-2-setup-an-etl-script">Step 2: Setup an ETL script</h3>

<p>Here we will outline the skeleton of an ETL script which utilizes the <code class="language-plaintext highlighter-rouge">c2m2_frictionless</code> package.</p>

<p>extract_transform.py:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#!/usr/bin/env python3
</span>
<span class="c1"># Import C2M2 Frictionless Helper package
## dataclasses
</span><span class="kn">import</span> <span class="nn">c2m2_frictionless</span>
<span class="kn">from</span> <span class="nn">c2m2_frictionless.C2M2_Level_1</span> <span class="kn">import</span> <span class="n">table_schema_specs_for_level_1_c2m2_encoding_of_dcc_metadata</span> <span class="k">as</span> <span class="n">c2m2_level_1</span>

<span class="c1"># TODO: Import other helper methods/setup DCC API
</span>
<span class="k">def</span> <span class="nf">extract_transform</span><span class="p">():</span>
  <span class="s">''' Generate c2m2 level 1 dataclasses to be added to the datapackage
  '''</span>
  <span class="c1"># Construct the core id_namespace separating this DCC's ids from the ids of
</span>  <span class="c1">#  all other DCCs
</span>  <span class="n">ns</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">(</span>
    <span class="c1"># Using a url is recommended
</span>    <span class="c1"># it is ideal if id_namespace + id actually resolves to a valid landing page
</span>    <span class="nb">id</span><span class="o">=</span><span class="s">'http://www.lincsproject.org/'</span><span class="p">,</span>
    <span class="n">abbreviation</span><span class="o">=</span><span class="s">'LINCS'</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s">'LINCS	Library of Integrated Network-Based Cellular Signatures'</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s">'A library that catalogs changes that occur when different types of cells are exposed to a variety of agents that disrupt normal cellular functions'</span><span class="p">,</span>
  <span class="p">)</span>
  <span class="k">yield</span> <span class="n">ns</span>
  <span class="c1"># TODO: yield additional models
</span>
<span class="k">def</span> <span class="nf">extract_transform_validate</span><span class="p">(</span><span class="n">outdir</span><span class="p">):</span>
  <span class="c1"># use the extract_transform generator to produce a datapackage
</span>  <span class="n">pkg</span> <span class="o">=</span> <span class="n">c2m2_frictionless</span><span class="o">.</span><span class="n">create_datapackage</span><span class="p">(</span><span class="s">'C2M2_Level_1'</span><span class="p">,</span> <span class="n">extract_transform</span><span class="p">(),</span> <span class="n">outdir</span><span class="p">)</span>
  <span class="c1"># identify ontological terms in the datapackage, and complete the term tables
</span>  <span class="n">c2m2_frictionless</span><span class="o">.</span><span class="n">build_term_tables</span><span class="p">(</span><span class="n">outdir</span><span class="p">)</span>
  <span class="c1"># validate assertion that the frictionless tableschema is complied with
</span>  <span class="n">c2m2_frictionless</span><span class="o">.</span><span class="n">validate_datapackage</span><span class="p">(</span><span class="n">pkg</span><span class="p">)</span>
  <span class="c1"># validate assertion that (id_namespace, name) is unique in each resource
</span>  <span class="n">c2m2_frictionless</span><span class="o">.</span><span class="n">validate_id_namespace_name_uniqueness</span><span class="p">(</span><span class="n">pkg</span><span class="p">)</span>
  <span class="c1">#
</span>  <span class="k">return</span> <span class="n">pkg</span>

<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">'__main__'</span><span class="p">:</span>
  <span class="kn">import</span> <span class="nn">sys</span>
  <span class="c1"># grab an output directory off the command line
</span>  <span class="n">extract_transform_validate</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span> <span class="k">else</span> <span class="s">''</span><span class="p">)</span>

</code></pre></div></div>

<p>With the above skeleton in place, we can already run this script <code class="language-plaintext highlighter-rouge">python3 etl.py output</code> to produce and validate a C2M2 Level 1 datapackage, though this package will only contain one element: the <code class="language-plaintext highlighter-rouge">id_namespace</code>. As described in the comments, this should be used to make sure that the ids within your DCC will not clash with the ids in another, for more about this please refer to the C2M2 <a href="https://cfde-published-documentation.readthedocs-hosted.com/en/latest/spec-and-docs/C2M2-usage-guides-and-technical-documents/000-INTRODUCTION/">documentation</a>.</p>

<p>Now we are ready to expand the <code class="language-plaintext highlighter-rouge">extract_transform</code> function such that we walk through the DCC’s pertinent dataset descriptions and yield C2M2 equivalent dataclasses.</p>

<h3 id="step-3-extracting-and-transforming-your-metadata-for-the-c2m2">Step 3: Extracting and transforming your metadata for the C2M2</h3>

<h4 id="step-31-projects">Step 3.1: Projects</h4>

<p>We will first focus on grabbing the <code class="language-plaintext highlighter-rouge">project</code> structure. It is recommended that a primary project for which all other projects point to is used for your DCC for logical grouping in the UI. Please see the <a href="https://cfde-published-documentation.readthedocs-hosted.com/en/latest/spec-and-docs/C2M2-usage-guides-and-technical-documents/000-INTRODUCTION/">C2M2 documentation</a> for a full description of what a “project” should represent, in short it is the groupings of datasets based essentially on funding awards.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">lincs_fetchdata_iter</span><span class="p">():</span>
  <span class="s">''' This method paginates through the LINCS API yielding LINCS datasets
  '''</span>
  <span class="c1"># ...
</span>
<span class="c1"># ...
</span><span class="k">def</span> <span class="nf">extract_transform</span><span class="p">():</span>
  <span class="c1"># ...
</span>  <span class="c1"># Projects refer to collections of items that were produced i.e. under the same award
</span>  <span class="c1">#  for a more complete definition of this and other C2M2 constructs,
</span>  <span class="c1">#  please refer to the C2M2 docs (https://www.nih-cfde.org/product/cfde-c2m2/)
</span>  <span class="n">primary_project</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">project</span><span class="p">(</span>
    <span class="c1"># The namespace from before, ensuring any ids we choose don't clash
</span>    <span class="n">id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
    <span class="nb">id</span><span class="o">=</span><span class="s">'lincs'</span><span class="p">,</span>
    <span class="c1"># the persistent id here enables you to point to a specific id,
</span>    <span class="c1">#  ideally a resolvable one, that behaves as a landing page for the project
</span>    <span class="n">persistent_id</span><span class="o">=</span><span class="s">'http://www.lincsproject.org/'</span><span class="p">,</span>
    <span class="c1"># this abbreviation may be used in the display of your project
</span>    <span class="n">abbreviation</span><span class="o">=</span><span class="s">'LINCS'</span><span class="p">,</span>
    <span class="c1"># this is the full name of your project
</span>    <span class="n">name</span><span class="o">=</span><span class="s">'Library of Integrated Network-based Cellular Signatures'</span><span class="p">,</span>
    <span class="c1"># this is the full description of your project
</span>    <span class="n">description</span><span class="o">=</span><span class="s">'The Library of Integrated Network-Based Cellular Signatures (LINCS) Program aims to create a network-based understanding of biology by cataloging changes in gene expression and other cellular processes that occur when cells are exposed to a variety of perturbing agents.'</span><span class="p">,</span>
  <span class="p">)</span>
  <span class="k">yield</span> <span class="n">primary_project</span>
  <span class="c1"># ...
</span>  <span class="n">projects</span> <span class="o">=</span> <span class="p">{}</span>
  <span class="c1"># ...
</span>  <span class="k">for</span> <span class="n">dataset</span> <span class="ow">in</span> <span class="n">lincs_fetchdata_iter</span><span class="p">():</span>
    <span class="c1"># LINCS datasets each refer to a `project`
</span>    <span class="c1">#  LINCS phase 1, 2, MCF10A, and more, these also fit as C2M2 project
</span>    <span class="c1">#
</span>    <span class="c1"># We will register the project if not done so already ensuring we
</span>    <span class="c1">#  don't end up with duplicates
</span>    <span class="k">if</span> <span class="n">dataset</span><span class="p">[</span><span class="s">'projectname'</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">projects</span><span class="p">:</span>
      <span class="n">project</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">project</span><span class="p">(</span>
        <span class="c1"># we'll use the same namespace of our primary project
</span>        <span class="n">id_namespace</span><span class="o">=</span><span class="n">primary_project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="c1"># we'll use the project name as the identifier with lack of a better
</span>        <span class="c1">#  solution
</span>        <span class="nb">id</span><span class="o">=</span><span class="n">dataset</span><span class="p">[</span><span class="s">'projectname'</span><span class="p">],</span>
        <span class="n">name</span><span class="o">=</span><span class="n">dataset</span><span class="p">[</span><span class="s">'projectname'</span><span class="p">],</span>
        <span class="c1"># we've omitted the optional persistent_id here because we currently
</span>        <span class="c1">#  don't have landing pages for these project names and as such
</span>        <span class="c1">#  cannot produce a persistent id referring to them. Ideally however
</span>        <span class="c1">#  we would use those urls.
</span>      <span class="p">)</span>
      <span class="c1"># we'll save the project for de-duplication purposes
</span>      <span class="n">projects</span><span class="p">[</span><span class="n">dataset</span><span class="p">[</span><span class="s">'projectname'</span><span class="p">]]</span> <span class="o">=</span> <span class="n">project</span>
      <span class="c1"># we'll yield it so it gets included in our datapackage
</span>      <span class="k">yield</span> <span class="n">project</span>
      <span class="c1"># we'll also create a project_in_project associating our primary project
</span>      <span class="c1">#  to the project we just made
</span>      <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">project_in_project</span><span class="p">(</span>
        <span class="n">parent_project_id_namespace</span><span class="o">=</span><span class="n">primary_project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="n">parent_project_id</span><span class="o">=</span><span class="n">primary_project</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
        <span class="n">child_project_id_namespace</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="n">child_project_id</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="c1"># grab the already created project
</span>      <span class="n">project</span> <span class="o">=</span> <span class="n">projects</span><span class="p">[</span><span class="n">dataset</span><span class="p">[</span><span class="s">'projectname'</span><span class="p">]]</span>
    <span class="c1"># ...
</span>  <span class="c1"># ...
</span>
</code></pre></div></div>

<h4 id="step-32-collections--files">Step 3.2: Collections &amp; Files</h4>

<p>With our project structure in place, we are now ready to walk through bringing the datasets and files into the package.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># ...
</span><span class="k">def</span> <span class="nf">extract_transform</span><span class="p">():</span>
  <span class="c1"># ...
</span>  <span class="k">for</span> <span class="n">dataset</span> <span class="ow">in</span> <span class="n">lincs_fetchdata_iter</span><span class="p">():</span>
    <span class="c1"># ...
</span>    <span class="c1"># grab creation time as iso 8601
</span>    <span class="n">creation_time</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'datereleased'</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">creation_time</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
      <span class="n">creation_time</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">strptime</span><span class="p">(</span>
        <span class="n">creation_time</span><span class="p">,</span> <span class="s">'</span><span class="si">%</span><span class="s">Y-</span><span class="si">%</span><span class="s">m-</span><span class="si">%</span><span class="s">d'</span>
      <span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
        <span class="n">tzinfo</span><span class="o">=</span><span class="n">datetime</span><span class="o">.</span><span class="n">timezone</span><span class="o">.</span><span class="n">utc</span>
      <span class="p">)</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()</span>
    <span class="c1"># each LINCS dataset group has multiple datasets,
</span>    <span class="c1">#  each LINCS dataset has multiple files.
</span>    <span class="c1"># We can model this with two collections referring to the independent
</span>    <span class="c1">#  sets of files
</span>    <span class="c1">#
</span>    <span class="c1"># LINCS Dataset Group
</span>    <span class="n">collection</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">collection</span><span class="p">(</span>
      <span class="n">id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="nb">id</span><span class="o">=</span><span class="n">dataset</span><span class="p">[</span><span class="s">'datasetgroup'</span><span class="p">],</span>
      <span class="n">persistent_id</span><span class="o">=</span><span class="n">f</span><span class="s">"http://lincsportal.ccs.miami.edu/datasets/view/{dataset['datasetgroup']}"</span><span class="p">,</span>
      <span class="n">name</span><span class="o">=</span><span class="n">dataset</span><span class="p">[</span><span class="s">'datasetname'</span><span class="p">],</span>
      <span class="n">description</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'description'</span><span class="p">,</span> <span class="s">''</span><span class="p">),</span>
      <span class="n">creation_time</span><span class="o">=</span><span class="n">creation_time</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">yield</span> <span class="n">collection</span>
    <span class="c1"># we can associate our collection with our project
</span>    <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">collection_defined_by_project</span><span class="p">(</span>
      <span class="n">collection_id_namespace</span><span class="o">=</span><span class="n">collection</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
      <span class="n">collection_id</span><span class="o">=</span><span class="n">collection</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="n">project_id_namespace</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
      <span class="n">project_id</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1">#
</span>    <span class="c1"># each dataset level (corresponding to a file) is specified in the metadata
</span>    <span class="k">for</span> <span class="n">dataset_id</span><span class="p">,</span> <span class="n">dataset_version</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'datasetlevels'</span><span class="p">,</span> <span class="p">[]),</span> <span class="n">dataset</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'latestversions'</span><span class="p">,</span> <span class="p">[])):</span>
      <span class="nb">file</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="nb">file</span><span class="p">(</span>
        <span class="n">id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
        <span class="c1"># the LINCS dataset id is used as the id
</span>        <span class="nb">id</span><span class="o">=</span><span class="n">dataset_id</span><span class="p">,</span>
        <span class="c1"># The file is necessarily associated with the project
</span>        <span class="n">project_id_namespace</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="n">project</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
        <span class="c1"># The landing page of the dataset is used as the persistent id
</span>        <span class="n">persistent_id</span><span class="o">=</span><span class="n">f</span><span class="s">"http://lincsportal.ccs.miami.edu/datasets/view/{dataset_id}"</span><span class="p">,</span>
        <span class="c1"># the creation time we converted to ISO8601 before
</span>        <span class="n">creation_time</span><span class="o">=</span><span class="n">creation_time</span><span class="p">,</span>
        <span class="c1"># File filename, note that C2M2 Level 1 doesn't currently have a distinction between
</span>        <span class="c1">#  download and landing page, we opted to put the download url in the filename.
</span>        <span class="n">filename</span><span class="o">=</span><span class="n">f</span><span class="s">"http://lincsportal.ccs.miami.edu/dcic/api/download?path=LINCS_Data/{datum['centername']}&amp;amp;file={dataset_id}_{dataset_version}.tar.gz"</span><span class="p">,</span>
        <span class="c1"># Following are additional optional fields that we were unable
</span>        <span class="c1">#  to easily map with LINCS but should be filled in if possible.
</span>        <span class="c1"># size_in_bytes=0,
</span>        <span class="c1"># sha256='',
</span>        <span class="c1"># md5='',
</span>        <span class="c1"># file_format='', # EDAM format:
</span>        <span class="c1"># data_type='',   # EDAM data:
</span>      <span class="p">)</span>
      <span class="k">yield</span> <span class="nb">file</span>
      <span class="c1"># each file is in the collection
</span>      <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">file_in_collection</span><span class="p">(</span>
        <span class="n">file_id_namespace</span><span class="o">=</span><span class="nb">file</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="n">file_id</span><span class="o">=</span><span class="nb">file</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
        <span class="n">collection_id_namespace</span><span class="o">=</span><span class="n">collection</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="n">collection_id</span><span class="o">=</span><span class="n">collection</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="p">)</span>

</code></pre></div></div>

<p>With this, we have already satisfied a C2M2 Level 0 instance, but ideally we’ll go a step further to C2M2 Level 1 which has more annotations including information about Subjects and Biosamples, for more information refer to the <a href="https://cfde-published-documentation.readthedocs-hosted.com/en/latest/spec-and-docs/C2M2-usage-guides-and-technical-documents/000-INTRODUCTION/#c2m2-richness-levels">C2M2 Level documentation</a>.</p>

<h4 id="step-33-subjects--biosamples">Step 3.3: Subjects &amp; Biosamples</h4>

<p>The LINCS data associates Datasets with small molecules, cell lines, genes, proteins, antibodies, and other entities related to experimental conditions and readouts, a dataset consists of a series of experiments mostly carried out in high throughput with a goal of capturing changes in expression or other phenotypic changes that may be captured via images in various contexts under various conditions (for example, perturbed by small molecules in different cell lines).</p>

<p>Due to the sheer number of experiments performed in high throughput, LINCS stores the whole set of experiments in a few files representing the data level (level of processing performed on the raw data). It might be possible to extract these files and produce massive amounts of biosamples, but this information is not currently directly available on the website for the practical reason that having landing pages for each micro-experiment is not useful for the users of this data.</p>

<p>Because of this, LINCS lumps all the biosamples together into one “biosample” referring to the whole experiment, this allows annotating that biosample with the assay used to perform the experiment and associate that biosample with the subjects of the experiment (cell line, small molecules, etc..) that were studied. This closely matches how the LINCS portal serves this data and fits in the C2M2.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#
</span><span class="k">def</span> <span class="nf">lincs_fetchstats</span><span class="p">(</span><span class="n">datasetid</span><span class="p">,</span> <span class="n">fields</span><span class="p">):</span>
  <span class="c1"># ...
#
# This dictionary describes how to get metadata for a given "stattype" (subject)
</span><span class="n">statstypes</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s">'protein'</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="s">'StatType'</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(),</span> <span class="nb">dict</span><span class="p">(</span>
    <span class="c1"># this describes the "type of subject"
</span>    <span class="nb">id</span><span class="o">=</span><span class="s">'lincs_subject_granularity:protein'</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s">'protein'</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s">'A protein biological entity'</span><span class="p">,</span>
    <span class="c1"># this is the landing page for this subject
</span>    <span class="n">download_url</span><span class="o">=</span><span class="k">lambda</span> <span class="n">datasetid</span><span class="p">:</span> <span class="n">f</span><span class="s">"http://lincsportal.ccs.miami.edu/dcic/api/Results?searchTerm=</span><span class="si">%22</span><span class="s">{datasetid}</span><span class="si">%22</span><span class="s">&amp;category=Protein"</span><span class="p">,</span>
    <span class="c1"># this is the id we should use for this subject
</span>    <span class="n">id_from_meta</span><span class="o">=</span><span class="k">lambda</span> <span class="n">meta</span><span class="p">:</span> <span class="n">meta</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'PR_UniProt_ID'</span><span class="p">)</span> <span class="ow">or</span> <span class="n">meta</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'PP_Name'</span><span class="p">),</span>
  <span class="p">)),</span>
  <span class="c1"># ...
</span><span class="p">}</span>
<span class="c1"># ...
</span><span class="k">def</span> <span class="nf">extract_transform</span><span class="p">():</span>
  <span class="c1"># ...
</span>  <span class="c1"># Each "stattype" corresponds to a "subject granularity" in the C2M2, we'll
</span>  <span class="c1">#  register those subject granularities now
</span>  <span class="n">subject_granularities</span> <span class="o">=</span> <span class="p">{}</span>
  <span class="c1"># Add all subject granularities from `stattypes`
</span>  <span class="k">for</span> <span class="n">stat_key</span><span class="p">,</span> <span class="n">stat</span> <span class="ow">in</span> <span class="n">statstypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">subject_granularities</span><span class="p">[</span><span class="n">stat_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">subject_granularity</span><span class="p">(</span>
      <span class="nb">id</span><span class="o">=</span><span class="n">stat</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="n">name</span><span class="o">=</span><span class="n">stat</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
      <span class="n">description</span><span class="o">=</span><span class="n">stat</span><span class="o">.</span><span class="n">description</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">yield</span> <span class="n">subject_granularities</span><span class="p">[</span><span class="n">stat_key</span><span class="p">]</span>
  <span class="c1"># ...
</span>  <span class="n">subjects</span> <span class="o">=</span> <span class="p">{}</span>
  <span class="c1"># ...
</span>  <span class="k">for</span> <span class="n">dataset</span> <span class="ow">in</span> <span class="n">lincs_fetchdata_iter</span><span class="p">():</span>
    <span class="c1"># ...
</span>    <span class="c1"># we'll associate a single biosample for the purpose of listing assay
</span>    <span class="n">biosample_id</span> <span class="o">=</span> <span class="n">collection</span><span class="o">.</span><span class="nb">id</span>
    <span class="c1"># NOTE: We defer biosample creation till after we have the anatomy
</span>    <span class="c1"># the biosample is in the collection
</span>    <span class="c1">#
</span>    <span class="c1"># a dataset_group has a number of aspects that are associated with it
</span>    <span class="c1"># including small molecules, proteins, cell lines, etc.. these are the subjects of the
</span>    <span class="c1"># experiment. We will create all of these and associate them with our single 'super' biosample
</span>    <span class="n">dataset_group_subjects</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">all_stats</span> <span class="o">=</span> <span class="n">lincs_fetchstats</span><span class="p">(</span><span class="n">datum</span><span class="p">[</span><span class="s">'datasetid'</span><span class="p">],</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">datum</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'statsfields'</span><span class="p">,</span> <span class="p">[])))</span>
    <span class="k">for</span> <span class="n">stat_name</span><span class="p">,</span> <span class="n">stats</span> <span class="ow">in</span> <span class="n">all_stats</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
      <span class="k">for</span> <span class="n">stat</span> <span class="ow">in</span> <span class="n">stats</span><span class="p">:</span>
        <span class="n">subject</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">subject_id</span> <span class="o">=</span> <span class="n">statstypes</span><span class="p">[</span><span class="n">stat_name</span><span class="p">]</span><span class="o">.</span><span class="n">id_from_meta</span><span class="p">(</span><span class="n">stat</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">subject_id</span><span class="p">:</span>
          <span class="k">if</span> <span class="n">stat_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">subjects</span><span class="p">:</span>
            <span class="n">subjects</span><span class="p">[</span><span class="n">stat_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
          <span class="k">if</span> <span class="n">subject_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">subjects</span><span class="p">[</span><span class="n">stat_name</span><span class="p">]:</span>
            <span class="n">subject</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">subject</span><span class="p">(</span>
              <span class="n">id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
              <span class="c1"># in an effort to ensure id uniqueness, we'll use the stat_name as a prefix
</span>              <span class="nb">id</span><span class="o">=</span><span class="n">f</span><span class="s">"{stat_name}:{subject_id}"</span><span class="p">,</span>
              <span class="c1"># our subject must be a part of the project, but because these subjects
</span>              <span class="c1">#   are re-used across all projects, we'll associate them with
</span>              <span class="c1">#   the primary project instead of just the project
</span>              <span class="n">project_id_namespace</span><span class="o">=</span><span class="n">primary_project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
              <span class="n">project</span><span class="o">=</span><span class="n">primary_project</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
              <span class="c1"># here we associate this subject with the subject granularity
</span>              <span class="n">granularity</span><span class="o">=</span><span class="n">subject_granularities</span><span class="p">[</span><span class="n">stat_name</span><span class="p">]</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">yield</span> <span class="n">subject</span>
            <span class="n">subjects</span><span class="p">[</span><span class="n">stat_name</span><span class="p">][</span><span class="n">subject_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">subject</span>
          <span class="k">else</span><span class="p">:</span>
            <span class="n">subject</span> <span class="o">=</span> <span class="n">subjects</span><span class="p">[</span><span class="n">stat_name</span><span class="p">][</span><span class="n">subject_id</span><span class="p">]</span>
          <span class="c1">#
</span>        <span class="c1"># if we were able to get a subject out of this we'll add it as an
</span>        <span class="c1">#  association to the biosample
</span>        <span class="k">if</span> <span class="n">subject</span> <span class="ow">and</span> <span class="n">subject</span><span class="o">.</span><span class="nb">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dataset_group_subjects</span><span class="p">:</span>
          <span class="n">dataset_group_subjects</span><span class="p">[</span><span class="n">subject</span><span class="o">.</span><span class="nb">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">subject</span>
          <span class="c1"># the biosample was derived from this subject
</span>          <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">biosample_from_subject</span><span class="p">(</span>
            <span class="n">biosample_id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
            <span class="n">biosample_id</span><span class="o">=</span><span class="n">biosample_id</span><span class="p">,</span>
            <span class="n">subject_id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
            <span class="n">subject_id</span><span class="o">=</span><span class="n">subject</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
          <span class="p">)</span>
          <span class="c1"># the subject is in the collection (lincs dataset group)
</span>          <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">subject_in_collection</span><span class="p">(</span>
            <span class="n">subject_id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
            <span class="n">subject_id</span><span class="o">=</span><span class="n">subject</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
            <span class="n">collection_id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
            <span class="n">collection_id</span><span class="o">=</span><span class="n">collection</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
          <span class="p">)</span>
    <span class="c1">#
</span>    <span class="n">biosample</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">biosample</span><span class="p">(</span>
      <span class="n">id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="nb">id</span><span class="o">=</span><span class="n">biosample_id</span><span class="p">,</span>
      <span class="n">project_id_namespace</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
      <span class="n">project</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="c1"># TODO: ontological mappings
</span>      <span class="c1"># assay_type='',
</span>      <span class="c1"># anatomy=''
</span>    <span class="p">)</span>
    <span class="k">yield</span> <span class="n">biosample</span>
    <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">biosample_in_collection</span><span class="p">(</span>
      <span class="n">biosample_id_namespace</span><span class="o">=</span><span class="n">biosample</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
      <span class="n">biosample_id</span><span class="o">=</span><span class="n">biosample_id</span><span class="p">,</span>
      <span class="n">collection_id_namespace</span><span class="o">=</span><span class="n">collection</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
      <span class="n">collection_id</span><span class="o">=</span><span class="n">collection</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># ...
</span>    <span class="k">for</span> <span class="n">dataset_id</span><span class="p">,</span> <span class="n">dataset_version</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">datum</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'datasetlevels'</span><span class="p">,</span> <span class="p">[]),</span> <span class="n">datum</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'latestversions'</span><span class="p">,</span> <span class="p">[])):</span>
      <span class="c1"># ...
</span>      <span class="c1"># each file describes the biosample
</span>      <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">file_describes_biosample</span><span class="p">(</span>
        <span class="n">file_id_namespace</span><span class="o">=</span><span class="nb">file</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="n">file_id</span><span class="o">=</span><span class="nb">file</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
        <span class="n">biosample_id_namespace</span><span class="o">=</span><span class="n">biosample</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
        <span class="n">biosample_id</span><span class="o">=</span><span class="n">biosample</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="p">)</span>
      <span class="c1"># each file also describes all of the subjects in the study
</span>      <span class="k">for</span> <span class="n">subject</span> <span class="ow">in</span> <span class="n">dataset_group_subjects</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">file_describes_subject</span><span class="p">(</span>
          <span class="n">file_id_namespace</span><span class="o">=</span><span class="nb">file</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
          <span class="n">file_id</span><span class="o">=</span><span class="nb">file</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
          <span class="n">subject_id_namespace</span><span class="o">=</span><span class="n">subject</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
          <span class="n">subject_id</span><span class="o">=</span><span class="n">subject</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="c1"># ...
</span></code></pre></div></div>

<h4 id="step-34-ontology-mapping">Step 3.4: Ontology mapping</h4>

<p>Without subjects and biosamples in place, we can now add ontological terms which will be a point of harmonization between the different DCCs. For some, this will be simple, especially if the CFDE is prescribing an ontology that is already being used. For others, this may be a more complicated mapping effort. In LINCS, assays were described in a much more specific manner than OBI, the currently prescribed ontology for assays in the C2M2–as such we need to collapse many of the assay descriptions into broader terms.</p>

<p>On the other hand, the organs were specified on the cell line in raw text, so these were mapped to the UBERON ontology.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># ...
</span><span class="n">lincs_assay_obi_lookup</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s">'Aggregated small molecule biochemical target activity'</span><span class="p">:</span> <span class="s">'OBI:0000070'</span><span class="p">,</span>
  <span class="s">'ATAC-seq'</span><span class="p">:</span> <span class="s">'OBI:0002020'</span><span class="p">,</span>
  <span class="s">'Bead-based immunoassay'</span><span class="p">:</span> <span class="s">'OBI:0001632'</span><span class="p">,</span>
  <span class="s">'ELISA'</span><span class="p">:</span> <span class="s">'OBI:0001632'</span><span class="p">,</span>
  <span class="s">'Fluorescence imaging'</span><span class="p">:</span> <span class="s">'OBI:0001501'</span><span class="p">,</span>
  <span class="s">'KiNativ'</span><span class="p">:</span> <span class="s">'OBI:0001146'</span><span class="p">,</span>
  <span class="s">'KINOMEscan'</span><span class="p">:</span> <span class="s">'OBI:0001146'</span><span class="p">,</span>
  <span class="s">'L1000'</span><span class="p">:</span> <span class="s">'OBI:0000424'</span><span class="p">,</span>
  <span class="s">'LC-MS'</span><span class="p">:</span> <span class="s">'OBI:0000470'</span><span class="p">,</span>
  <span class="s">'Mass spectrometry'</span><span class="p">:</span> <span class="s">'OBI:0000470'</span><span class="p">,</span>
  <span class="s">'MEMA (fluorescence imaging)'</span><span class="p">:</span> <span class="s">'OBI:0001501'</span><span class="p">,</span>
  <span class="s">'Microwestern'</span><span class="p">:</span> <span class="s">'OBI:0000615'</span><span class="p">,</span>
  <span class="s">'P100 (mass spectrometry)'</span><span class="p">:</span> <span class="s">'OBI:0000615'</span><span class="p">,</span>
  <span class="s">'PCR'</span><span class="p">:</span> <span class="s">'OBI:0001146'</span><span class="p">,</span>
  <span class="s">'RNA-seq'</span><span class="p">:</span> <span class="s">'OBI:0001271'</span><span class="p">,</span>
  <span class="s">'RPPA'</span><span class="p">:</span> <span class="s">'OBI:0000615'</span><span class="p">,</span>
  <span class="s">'SWATH-MS'</span><span class="p">:</span> <span class="s">'OBI:0000615'</span><span class="p">,</span>
  <span class="s">'Tandem Mass Tag (TMT) Mass Spectrometry'</span><span class="p">:</span> <span class="s">'OBI:0000470'</span><span class="p">,</span>
  <span class="s">'To Be Classified'</span><span class="p">:</span> <span class="s">''</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">lincs_anatomy_lookup</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s">'Ovary'</span><span class="p">:</span> <span class="s">'UBERON:0000992'</span><span class="p">,</span>
  <span class="s">'cervix'</span><span class="p">:</span> <span class="s">'UBERON:0000002'</span><span class="p">,</span>
  <span class="s">'kidney'</span><span class="p">:</span> <span class="s">'UBERON:0002113'</span><span class="p">,</span>
  <span class="s">'vulva'</span><span class="p">:</span> <span class="s">'UBERON:0000997'</span><span class="p">,</span>
  <span class="s">'gall bladder'</span><span class="p">:</span> <span class="s">'UBERON:0002110'</span><span class="p">,</span>
  <span class="s">'urinary bladder'</span><span class="p">:</span> <span class="s">'UBERON:0001255'</span><span class="p">,</span>
  <span class="s">'nervous system'</span><span class="p">:</span> <span class="s">'UBERON:0001016'</span><span class="p">,</span>
  <span class="s">'Lung'</span><span class="p">:</span> <span class="s">'UBERON:0002048'</span><span class="p">,</span>
  <span class="s">'skin'</span><span class="p">:</span> <span class="s">'UBERON:0002097'</span><span class="p">,</span>
  <span class="s">'Intestine'</span><span class="p">:</span> <span class="s">'UBERON:0000160'</span><span class="p">,</span>
  <span class="s">'Peripheral blood'</span><span class="p">:</span> <span class="s">'UBERON:0000178'</span><span class="p">,</span> <span class="c1"># WARNING: not 'peripheral'
</span>  <span class="s">'miscellaneous'</span><span class="p">:</span> <span class="s">''</span><span class="p">,</span>
  <span class="s">'adrenal gland'</span><span class="p">:</span> <span class="s">'UBERON:0002369'</span><span class="p">,</span>
  <span class="s">'lung'</span><span class="p">:</span> <span class="s">'UBERON:0002048'</span><span class="p">,</span>
  <span class="s">'thyroid'</span><span class="p">:</span> <span class="s">'UBERON:0002046'</span><span class="p">,</span>
  <span class="s">'urinary tract'</span><span class="p">:</span> <span class="s">'UBERON:0011143, UBERON:0001556'</span><span class="p">,</span>
  <span class="s">'uterus'</span><span class="p">:</span> <span class="s">'UBERON:0000995'</span><span class="p">,</span>
  <span class="s">'pancreas'</span><span class="p">:</span> <span class="s">'UBERON:0001264'</span><span class="p">,</span>
  <span class="s">'lymphatic system'</span><span class="p">:</span> <span class="s">'UBERON:0006558'</span><span class="p">,</span>
  <span class="s">'muscle'</span><span class="p">:</span> <span class="s">'UBERON:0001630'</span><span class="p">,</span>
  <span class="s">'biliary tract'</span><span class="p">:</span> <span class="s">''</span><span class="p">,</span>
  <span class="s">'blood'</span><span class="p">:</span> <span class="s">'UBERON:0000178'</span><span class="p">,</span>
  <span class="s">'stomach'</span><span class="p">:</span> <span class="s">'UBERON:0000945'</span><span class="p">,</span>
  <span class="s">'intestine'</span><span class="p">:</span> <span class="s">'UBERON:0000160'</span><span class="p">,</span>
  <span class="s">'head and neck'</span><span class="p">:</span> <span class="s">'UBERON:0000974'</span><span class="p">,</span> <span class="c1">#'UBERON:0000974, UBERON:0000974',
</span>  <span class="s">'human'</span><span class="p">:</span> <span class="s">''</span><span class="p">,</span>
  <span class="s">'Brain'</span><span class="p">:</span> <span class="s">'UBERON:0000955'</span><span class="p">,</span>
  <span class="s">'testes'</span><span class="p">:</span> <span class="s">'UBERON:0000473'</span><span class="p">,</span>
  <span class="s">'bone'</span><span class="p">:</span> <span class="s">'UBERON:0002481'</span><span class="p">,</span>
  <span class="s">'large intestine'</span><span class="p">:</span> <span class="s">'UBERON:0000059'</span><span class="p">,</span>
  <span class="s">'endometrium'</span><span class="p">:</span> <span class="s">'UBERON:0001295'</span><span class="p">,</span>
  <span class="s">'ureter'</span><span class="p">:</span> <span class="s">'UBERON:0000056'</span><span class="p">,</span>
  <span class="s">'pleura'</span><span class="p">:</span> <span class="s">'UBERON:0000977'</span><span class="p">,</span>
  <span class="s">'Skin'</span><span class="p">:</span> <span class="s">'UBERON:0002097'</span><span class="p">,</span>
  <span class="s">'breast'</span><span class="p">:</span> <span class="s">'UBERON:0000310'</span><span class="p">,</span>
  <span class="s">'liver'</span><span class="p">:</span> <span class="s">'UBERON:0002107'</span><span class="p">,</span>
  <span class="s">'Mammary gland'</span><span class="p">:</span> <span class="s">'UBERON:0001911'</span><span class="p">,</span>
  <span class="s">'ovary'</span><span class="p">:</span> <span class="s">'UBERON:0000992'</span><span class="p">,</span>
  <span class="s">'prostate'</span><span class="p">:</span> <span class="s">'UBERON:0002367'</span><span class="p">,</span>
  <span class="s">'esophagus'</span><span class="p">:</span> <span class="s">'UBERON:0001043'</span><span class="p">,</span>
  <span class="s">'brain'</span><span class="p">:</span> <span class="s">'UBERON:0000955'</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">lincs_organism_lookup</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s">'Homo sapiens'</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="s">'NCBI:taxid9606'</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(),</span> <span class="nb">dict</span><span class="p">(</span>
    <span class="nb">id</span><span class="o">=</span><span class="s">'NCBI:taxid9606'</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="s">'species'</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s">'Homo sapiens'</span><span class="p">,</span>
  <span class="p">)),</span>
  <span class="s">'Homo Sapiens'</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="s">'NCBI:taxid9606'</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(),</span> <span class="nb">dict</span><span class="p">(</span>
    <span class="nb">id</span><span class="o">=</span><span class="s">'NCBI:taxid9606'</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="s">'species'</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s">'Homo sapiens'</span><span class="p">,</span>
  <span class="p">)),</span>
  <span class="s">'Homo Sapien'</span><span class="p">:</span> <span class="nb">type</span><span class="p">(</span><span class="s">'NCBI:taxid9606'</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(),</span> <span class="nb">dict</span><span class="p">(</span>
    <span class="nb">id</span><span class="o">=</span><span class="s">'NCBI:taxid9606'</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="s">'species'</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s">'Homo sapiens'</span><span class="p">,</span>
  <span class="p">)),</span>
<span class="p">}</span>
<span class="c1"># ...
</span><span class="k">def</span> <span class="nf">first_or_none</span><span class="p">(</span><span class="n">it</span><span class="p">):</span>
  <span class="s">''' A simple helper for safely returning the first element of a potentially empty iterator
  '''</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="k">return</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">it</span><span class="p">))</span>
  <span class="k">except</span> <span class="nb">StopIteration</span><span class="p">:</span>
    <span class="k">return</span> <span class="bp">None</span>
<span class="c1"># ...
</span><span class="k">def</span> <span class="nf">extract_transform</span><span class="p">():</span>
  <span class="c1"># ...
</span>  <span class="n">taxonomies</span> <span class="o">=</span> <span class="p">{}</span>
  <span class="c1"># https://github.com/nih-cfde/specifications-and-documentation/blob/master/draft-C2M2_internal_CFDE_CV_tables/subject_granularity.tsv#L2
</span>  <span class="n">cfde_subject_role_organism</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">subject_role</span><span class="p">(</span>
    <span class="nb">id</span><span class="o">=</span><span class="s">'cfde_subject_role:0'</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s">'single organism'</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s">"The organism represented by a subject in the 'single organism' granularity category"</span><span class="p">,</span>
  <span class="p">)</span>
  <span class="k">yield</span> <span class="n">cfde_subject_role_organism</span>
  <span class="c1"># ...
</span>  <span class="k">for</span> <span class="n">dataset</span> <span class="ow">in</span> <span class="n">lincs_fetchdata_iter</span><span class="p">():</span>
    <span class="c1"># ...
</span>    <span class="n">anatomies</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="c1">#
</span>    <span class="k">for</span> <span class="n">stat_name</span><span class="p">,</span> <span class="n">stats</span> <span class="ow">in</span> <span class="n">all_stats</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
      <span class="c1"># ...
</span>      <span class="k">if</span> <span class="n">subject</span><span class="p">:</span>
        <span class="c1"># ...
</span>        <span class="c1"># In order to capture anatomy at the biosample level, we need to catch
</span>        <span class="c1">#  the anatomy on the subject
</span>        <span class="k">if</span> <span class="n">stat_name</span> <span class="o">==</span> <span class="s">'cellline'</span><span class="p">:</span>
          <span class="c1"># in the case of cell lines, we'd have a taxonomy
</span>          <span class="n">lincs_taxon</span> <span class="o">=</span> <span class="n">lincs_organism_lookup</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'organism'</span><span class="p">))</span>
          <span class="k">if</span> <span class="n">lincs_taxon</span> <span class="ow">and</span> <span class="n">lincs_taxon</span><span class="o">.</span><span class="nb">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxonomies</span><span class="p">:</span>
            <span class="n">taxon</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">ncbi_taxonomy</span><span class="p">(</span>
              <span class="nb">id</span><span class="o">=</span><span class="n">lincs_taxon</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
              <span class="n">clade</span><span class="o">=</span><span class="n">lincs_taxon</span><span class="o">.</span><span class="n">clade</span><span class="p">,</span>
              <span class="n">name</span><span class="o">=</span><span class="n">lincs_taxon</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">taxonomies</span><span class="p">[</span><span class="n">lincs_taxon</span><span class="o">.</span><span class="nb">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">taxon</span>
            <span class="k">yield</span> <span class="n">taxon</span>
          <span class="k">elif</span> <span class="n">lincs_taxon</span><span class="p">:</span>
            <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxonomies</span><span class="p">[</span><span class="n">lincs_taxon</span><span class="o">.</span><span class="nb">id</span><span class="p">]</span>
          <span class="k">else</span><span class="p">:</span>
            <span class="n">taxon</span> <span class="o">=</span> <span class="bp">None</span>
          <span class="c1">#
</span>          <span class="k">if</span> <span class="n">taxon</span><span class="p">:</span>
            <span class="c1"># associate the subject with the taxon
</span>            <span class="k">yield</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">subject_role_taxonomy</span><span class="p">(</span>
              <span class="n">subject_id_namespace</span><span class="o">=</span><span class="n">subject</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
              <span class="n">subject_id</span><span class="o">=</span><span class="n">subject</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
              <span class="n">role_id</span><span class="o">=</span><span class="n">subject_role_organism</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
              <span class="n">taxonomy_id</span><span class="o">=</span><span class="n">taxon</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
            <span class="p">)</span>
          <span class="c1"># in the case of cell lines, we'd have an anatomy
</span>          <span class="k">if</span> <span class="n">subject_id</span> <span class="ow">in</span> <span class="n">lincs_anatomy_lookup</span><span class="p">:</span>
            <span class="n">anatomies</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">lincs_cellline_anatomy</span><span class="p">(</span><span class="n">lincs_anatomy_lookup</span><span class="p">[</span><span class="n">subject_id</span><span class="p">]))</span>
          <span class="k">else</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="n">f</span><span class="s">"WARNING: {subject_id} not in celllines"</span><span class="p">)</span>
    <span class="c1"># ...
</span>    <span class="n">biosample</span> <span class="o">=</span> <span class="n">c2m2_level_1</span><span class="o">.</span><span class="n">biosample</span><span class="p">(</span>
      <span class="n">id_namespace</span><span class="o">=</span><span class="n">ns</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="nb">id</span><span class="o">=</span><span class="n">biosample_id</span><span class="p">,</span>
      <span class="n">project_id_namespace</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="n">id_namespace</span><span class="p">,</span>
      <span class="n">project</span><span class="o">=</span><span class="n">project</span><span class="o">.</span><span class="nb">id</span><span class="p">,</span>
      <span class="c1"># Lookup assay name to get OBI
</span>      <span class="n">assay_type</span><span class="o">=</span><span class="n">lincs_assay_obi_lookup</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">'assayname'</span><span class="p">),</span> <span class="s">''</span><span class="p">),</span>
      <span class="c1"># NOTE: we have several anatomies at this point and are forced to choose one
</span>      <span class="c1"># this will require improvements to either the model or to our conversion
</span>      <span class="c1"># perhaps we could instead have several biosamples for each file corresponding
</span>      <span class="c1"># to each subject. This will be left for future directions
</span>      <span class="n">anatomy</span><span class="o">=</span><span class="n">first_or_none</span><span class="p">(</span><span class="n">anatomies</span><span class="p">),</span>
    <span class="p">)</span>
</code></pre></div></div>

<p>Here we show how we were able to capture an anatomy, taxonomies on cell lines, and assays on biosamples. While not complete and not perfect, this provides the much needed interoperation layer with the other DCCs. With this type of annotation, we can look at files across different DCCs by assay, anatomy, or subjects by taxonomy. It improves the findability of these files even if the model does not perfectly capture the real structure of the original page. The CFDE is meant to catalog and provide federated search to all of the DCCs, directing users to the DCC’s pages, not to replace the catalogs already in place.</p>

<h3 id="step-4-running-and-validating-our-extraction-and-transformation">Step 4: Running and validating our extraction and transformation</h3>

<p>Our script is now ready to be run, tested, and refactored. During this phase it is ideal to cache network requests if hitting API endpoints. Every run of the script will re-construct the entire data package and run through the various checks, reporting any errors that occur.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>python3 extract_transform.py output
</code></pre></div></div>

<h3 id="step-5-pushing-our-validated-c2m2-datapackage">Step 5: Pushing our validated C2M2 datapackage</h3>

<p>A client for facilitating the ingestion of your data into the CFDE portal exists <a href="https://github.com/fair-research/deriva-flow-client/tree/client-dev">here</a>. Once installed, the output directory we produced with the <code class="language-plaintext highlighter-rouge">etl.py</code> script can be sent to DERIVA.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># install the DERIVA loader script</span>
pip <span class="nb">install </span>git+git://github.com/fair-research/deriva-flow-client.git@client-dev

<span class="c"># output here is the directory with the output of the etl.py script</span>
cfde run output

<span class="c"># check the status of the ingestion</span>
cfde status
</code></pre></div></div>

<p>This tool facilitates authentication, manual preview verification when loaded on the server side (you will get an email), followed by CFDE verification and ultimate incorporation of your data into the unified portal.</p>

<h2 id="conclusion">Conclusion</h2>

<p>Taking advantage of this tooling will simplify the process of ETL script development given that most things will be caught with dataclass assertions and datapackage validation. Nonetheless, this enforces minimal compliance with the C2M2 standard. For maximal compliance, it is essential that you review the most up to date <a href="https://cfde-published-documentation.readthedocs-hosted.com/en/latest/spec-and-docs/C2M2-usage-guides-and-technical-documents/000-INTRODUCTION/">C2M2 documentation</a> and it may also be useful to review and perform <a href="https://nih-cfde.github.io/the-fair-cookbook/recipes/04/fairshake.html">FAIR assessments</a> which include more elaborate assertions for compliance with FAIRness beyond the C2M2.</p>

            </div>
            <div class="c-textbook__footer" id="textbook_footer">
              
<nav class="c-page__nav">
  

  
</nav>

              <footer>
  <p class="footer">This project is supported by NIH CFDE program <a href="https://commonfund.nih.gov/dataecosystem"><img src="/the-fair-cookbook/assets/images/logo/CFDE-logo.png" height="20" alt="NIH flag" class="float-left bottom-margin"></a> <script>mermaid.init();</script></p>
</footer>

            </div>

        </div>
      </main>
    </div>
  </body>
</html>
