<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width,minimum-scale=1">

  <title>CFDE Glossary</title>
  <meta name="description" content="# CFDE Glossary## Controlled Vocabulary**Asset** - a sample or a file.**Biospecimens** - a material collected from an organism, a cell culture, or a material...">

  <link rel="canonical" href="https://nih-cfde.github.io/the-fair-cookbook/content/CFDE_Glossary.html">
  <link rel="alternate" type="application/rss+xml" title="NIH-CFDE FAIR COOKBOOK" href="https://nih-cfde.github.io/the-fair-cookbook/feed.xml">

  <meta property="og:url"         content="https://nih-cfde.github.io/the-fair-cookbook/content/CFDE_Glossary.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="CFDE Glossary" />
<meta property="og:description" content="# CFDE Glossary## Controlled Vocabulary**Asset** - a sample or a file.**Biospecimens** - a material collected from an organism, a cell culture, or a material..." />
<meta property="og:image"       content="https://nih-cfde.github.io/the-fair-cookbook/images/logo/CFDE-logo-contrast.png" />

<meta name="twitter:card" content="summary">


  <script type="application/ld+json">
  {
  "@context": "http://schema.org",
  "@type": "NewsArticle",
  "mainEntityOfPage": "https://nih-cfde.github.io/the-fair-cookbook/content/CFDE_Glossary.html",
  "headline": "CFDE Glossary",
  "datePublished": "2020-06-09T17:11:14+00:00",
  "dateModified": "2020-06-09T17:11:14+00:00",
  "description": "# CFDE Glossary## Controlled Vocabulary**Asset** - a sample or a file.**Biospecimens** - a material collected from an organism, a cell culture, or a material...",
  "author": {
    "@type": "Person",
    "name": "CFDE"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Data 100 at UC Berkeley",
    "logo": {
      "@type": "ImageObject",
      "url": "https://nih-cfde.github.io/the-fair-cookbook",
      "width": 60,
      "height": 60
    }
  },
  "image": {
    "@type": "ImageObject",
    "url": "https://nih-cfde.github.io/the-fair-cookbook",
    "height": 60,
    "width": 60
  }
}

  </script>
  <link rel="stylesheet" href="/the-fair-cookbook/assets/css/styles.css">

  <!-- <link rel="manifest" href="/manifest.json"> -->
  <!-- <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#efae0a"> -->
  <meta name="msapplication-TileColor" content="#da532c">
  <meta name="msapplication-TileImage" content="/mstile-144x144.png">
  <meta name="theme-color" content="#233947">

  <!-- Favicon -->
  <link rel="shortcut icon" type="image/x-icon" href="/the-fair-cookbook/images/logo/CFDE-logo.ico">

  <!-- MathJax Config -->
  <!-- Allow inline math using $ and automatically break long math lines -->
<!-- (mostly) copied from nbconvert configuration -->
<!-- https://github.com/jupyter/nbconvert/blob/master/nbconvert/templates/html/mathjax.tpl -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true },
    },
    
});
</script>
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS_HTML' async></script>


  <!-- DOM updating function -->
  <script src="/the-fair-cookbook/assets/js/page/dom-update.js"></script>

  <!-- Selectors for elements on the page -->
  <script src="/the-fair-cookbook/assets/js/page/documentSelectors.js"></script>

  <!-- Define some javascript variables that will be useful in other javascript -->
  <script>
    const site_basename = '/the-fair-cookbook';
  </script>

  <!-- Add AnchorJS to let headers be linked -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/4.2.0/anchor.min.js" async></script>
  <script src="/the-fair-cookbook/assets/js/page/anchors.js" async></script>

  <!-- Include Turbolinks to make page loads fast -->
  <!-- https://github.com/turbolinks/turbolinks -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/turbolinks/5.2.0/turbolinks.js" async></script>
  <meta name="turbolinks-cache-control" content="no-cache">

  <!-- Load nbinteract for widgets -->
  

  <!-- Load Thebelab for interactive widgets -->
  <!-- Include Thebelab for interactive code if it's enabled -->


<!-- Display Thebelab button in each code cell -->
<script>
/**
 * Set up thebelab button for code blocks
 */

const thebelabCellButton = id =>
  `<a id="thebelab-cell-button-${id}" class="btn thebebtn o-tooltip--left" data-tooltip="Interactive Mode">
    <img src="/the-fair-cookbook/assets/images/edit-button.svg" alt="Start thebelab interactive mode">
  </a>`


const addThebelabButtonToCodeCells =  () => {

  const codeCells = document.querySelectorAll('div.input_area > div.highlight:not(.output) pre')
  codeCells.forEach((codeCell, index) => {
    const id = codeCellId(index)
    codeCell.setAttribute('id', id)
    if (document.getElementById("thebelab-cell-button-" + id) == null) {
      codeCell.insertAdjacentHTML('afterend', thebelabCellButton(id));
    }
  })
}

initFunction(addThebelabButtonToCodeCells);
</script>


<script src="https://unpkg.com/thebelab@latest/lib/index.js" async></script>
<script>
    /**
     * Add attributes to Thebelab blocks
     */

    const initThebelab = () => {
        const addThebelabToCodeCells = () => {
            console.log("Adding thebelab to code cells...");
            // If Thebelab hasn't loaded, wait a bit and try again. This
            // happens because we load ClipboardJS asynchronously.
            if (window.thebelab === undefined) {
                setTimeout(addThebelabToCodeCells, 250)
            return
            }

            // If we already detect a Thebelab cell, don't re-run
            if (document.querySelectorAll('div.thebelab-cell').length > 0) {
                return;
            }

            // Find all code cells, replace with Thebelab interactive code cells
            const codeCells = document.querySelectorAll('.input_area pre')
            codeCells.forEach((codeCell, index) => {
                const id = codeCellId(index)

                // Clean up the language to make it work w/ CodeMirror and add it to the cell
                dataLanguage = ""
                dataLanguage = detectLanguage(dataLanguage);
                codeCell.setAttribute('data-language', dataLanguage)
                codeCell.setAttribute('data-executable', 'true')

                // If the code cell is hidden, show it
                var inputCheckbox = document.querySelector(`input#hidebtn${codeCell.id}`);
                if (inputCheckbox !== null) {
                    setCodeCellVisibility(inputCheckbox, 'visible');
                }
            });

            // Remove the event listener from the page so keyboard press doesn't
            // Change page
            document.removeEventListener('keydown', initPageNav)
            keyboardListener = false;

            // Init thebelab
            thebelab.bootstrap();

            // Remove copy buttons since they won't work anymore
            const copyAndThebeButtons = document.querySelectorAll('.copybtn, .thebebtn')
            copyAndThebeButtons.forEach((button, index) => {
                button.remove();
            });

            // Remove outputs since they'll be stale
            const outputs = document.querySelectorAll('.output *, .output')
            outputs.forEach((output, index) => {
                output.remove();
            });

            // Find any cells with an initialization tag and ask ThebeLab to run them when ready
            var thebeInitCells = document.querySelectorAll('div.tag_thebelab-init');
            thebeInitCells.forEach((cell) => {
                console.log("Initializing ThebeLab with cell: " + cell.id);
                cell.querySelector('.thebelab-run-button').click();
            });
        }

        // Add event listener for the function to modify code cells
        const thebelabButtons = document.querySelectorAll('[id^=thebelab], [id$=thebelab]')
        thebelabButtons.forEach((thebelabButton,index) => {
            if (thebelabButton === null) {
                setTimeout(initThebelab, 250)
                return
            };
            thebelabButton.addEventListener('click', addThebelabToCodeCells);
        });
    }

    // Initialize Thebelab
    initFunction(initThebelab);

// Helper function to munge the language name
var detectLanguage = (language) => {
    if (language.indexOf('python') > -1) {
        language = "python";
    }
    return language;
}
</script>



  <!-- Load the auto-generating TOC (non-async otherwise the TOC won't load w/ turbolinks) -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.8.1/tocbot.min.js" async></script>
  <script src="/the-fair-cookbook/assets/js/page/tocbot.js"></script>



<!-- using mermaid for rendering Diagram and flow chart -->
<!-- <script src="https://cdnjs.cloudflare.com/ajax/libs/mermaid/8.4.8/mermaid.min.js" async></script> -->
<script src="/the-fair-cookbook/assets/js/mermaid.min.js"></script>
<script>mermaid.initialize({startOnLoad:true});</script>
<!--  <script src="/the-fair-cookbook/assets/js/mermaid.js"></script>-->



  <!-- Google analytics -->
  
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id="></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '');
</script>



  <!-- Clipboard copy button -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" async></script>

  <!-- Load custom website scripts -->
  <script src="/the-fair-cookbook/assets/js/scripts.js" async></script>

  <!-- Load custom user CSS and JS  -->
  <script src="/the-fair-cookbook/assets/custom/custom.js" async></script>
  <link rel="stylesheet" href="/the-fair-cookbook/assets/custom/custom.css">

  <!-- Update interact links w/ REST param, is defined in includes so we can use templates -->
  
<script>
/**
  * To auto-embed hub URLs in interact links if given in a RESTful fashion
 */

function getJsonFromUrl(url) {
  var query = url.split('?');
  if (query.length < 2) {
    // No queries so just return false
    return false;
  }
  query = query[1];
  // Collect REST params into a dictionary
  var result = {};
  query.split("&").forEach(function(part) {
    var item = part.split("=");
    result[item[0]] = decodeURIComponent(item[1]);
  });
  return result;
}
    
function dict2param(dict) {
    params = Object.keys(dict).map(function(k) {
        return encodeURIComponent(k) + '=' + encodeURIComponent(dict[k])
    });
    return params.join('&')
}

// Parse a Binder URL, converting it to the string needed for JupyterHub
function binder2Jupyterhub(url) {
  newUrl = {};
  parts = url.split('v2/gh/')[1];
  // Grab the base repo information
  repoinfo = parts.split('?')[0];
  var [org, repo, ref] = repoinfo.split('/');
  newUrl['repo'] = ['https://github.com', org, repo].join('/');
  newUrl['branch'] = ref
  // Grab extra parameters passed
  params = getJsonFromUrl(url);
  if (params['filepath'] !== undefined) {
    newUrl['subPath'] = params['filepath']
  }
  return dict2param(newUrl);
}

// Filter out potentially unsafe characters to prevent xss
function safeUrl(url)
{
   return String(encodeURIComponent(url))
            .replace(/&/g, '&amp;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&#39;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;');
}

function addParamToInternalLinks(hub) {
  var links = document.querySelectorAll("a").forEach(function(link) {
    var href = link.href;
    // If the link is an internal link...
    if (href.search("https://nih-cfde.github.io") !== -1 || href.startsWith('/') || href.search("127.0.0.1:") !== -1) {
      // Assume we're an internal link, add the hub param to it
      var params = getJsonFromUrl(href);
      if (params !== false) {
        // We have REST params, so append a new one
        params['jupyterhub'] = hub;
      } else {
        // Create the REST params
        params = {'jupyterhub': hub};
      }
      // Update the link
      var newHref = href.split('?')[0] + '?' + dict2param(params);
      link.setAttribute('href', decodeURIComponent(newHref));
    }
  });
  return false;
}


// Update interact links
function updateInteractLink() {
    // hack to make this work since it expects a ? in the URL
    rest = getJsonFromUrl("?" + location.search.substr(1));
    jupyterHubUrl = rest['jupyterhub'];
    var hubType = null;
    var hubUrl = null;
    if (jupyterHubUrl !== undefined) {
      hubType = 'jupyterhub';
      hubUrl = jupyterHubUrl;
    }

    if (hubType !== null) {
      // Sanitize the hubUrl
      hubUrl = safeUrl(hubUrl);

      // Add HTTP text if omitted
      if (hubUrl.indexOf('http') < 0) {hubUrl = 'http://' + hubUrl;}
      var interactButtons = document.querySelectorAll("button.interact-button")
      var lastButton = interactButtons[interactButtons.length-1];
      var link = lastButton.parentElement;

      // If we've already run this, skip the link updating
      if (link.nextElementSibling !== null) {
        return;
      }

      // Update the link and add context div
      var href = link.getAttribute('href');
      if (lastButton.id === 'interact-button-binder') {
        // If binder links exist, we need to re-work them for jupyterhub
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // If localhost, assume we're working from a local Jupyter server and remove `/hub`
          first = [hubUrl, 'git-sync'].join('/')
        } else {
          first = [hubUrl, 'hub', 'user-redirect', 'git-sync'].join('/')
        }
        href = first + '?' + binder2Jupyterhub(href);
      } else {
        // If interact button isn't binderhub, assume it's jupyterhub
        // If JupyterHub links, we only need to replace the hub url
        href = href.replace("", hubUrl);
        if (hubUrl.indexOf('http%3A%2F%2Flocalhost') > -1) {
          // Assume we're working from a local Jupyter server and remove `/hub`
          href = href.replace("/hub/user-redirect", "");
        }
      }
      link.setAttribute('href', decodeURIComponent(href));

      // Add text after interact link saying where we're launching
      hubUrlNoHttp = decodeURIComponent(hubUrl).replace('http://', '').replace('https://', '');
      link.insertAdjacentHTML('afterend', '<div class="interact-context">on ' + hubUrlNoHttp + '</div>');

      // Update internal links so we retain the hub url
      addParamToInternalLinks(hubUrl);
    }
}

runWhenDOMLoaded(updateInteractLink)
document.addEventListener('turbolinks:load', updateInteractLink)
</script>


  <!-- Lunr search code - will only be executed on the /search page -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.6/lunr.min.js" async></script>
  <script>var initQuery = function() {
  // See if we have a search box
  var searchInput = document.querySelector('input#lunr_search');
  if (searchInput === null) {
    return;
  }

  // Function to parse our lunr cache
  var idx = lunr(function () {
    this.field('title')
    this.field('excerpt')
    this.field('categories')
    this.field('tags')
    this.ref('id')

    this.pipeline.remove(lunr.trimmer)

    for (var item in store) {
      this.add({
        title: store[item].title,
        excerpt: store[item].excerpt,
        categories: store[item].categories,
        tags: store[item].tags,
        id: item
      })
    }
  });

  // Run search upon keyup
  searchInput.addEventListener('keyup', function () {
    var resultdiv = document.querySelector('#results');
    var query = document.querySelector("input#lunr_search").value.toLowerCase();
    var result =
      idx.query(function (q) {
        query.split(lunr.tokenizer.separator).forEach(function (term) {
          q.term(term, { boost: 100 })
          if(query.lastIndexOf(" ") != query.length-1){
            q.term(term, {  usePipeline: false, wildcard: lunr.Query.wildcard.TRAILING, boost: 10 })
          }
          if (term != ""){
            q.term(term, {  usePipeline: false, editDistance: 1, boost: 1 })
          }
        })
      });

      // Empty the results div
      while (resultdiv.firstChild) {
        resultdiv.removeChild(resultdiv.firstChild);
      }

    resultdiv.insertAdjacentHTML('afterbegin', '<p class="results__found">'+result.length+' Result(s) found</p>');
    for (var item in result) {
      var ref = result[item].ref;
      if(store[ref].teaser){
        var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<div class="archive__item-teaser">'+
                '<img src="'+store[ref].teaser+'" alt="">'+
              '</div>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      else{
    	  var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      resultdiv.insertAdjacentHTML('beforeend', searchitem);
    }
  });
};

initFunction(initQuery);
</script>

  <!-- Load JS that depends on site variables -->
  <script src="/the-fair-cookbook/assets/js/page/copy-button.js" async></script>

  <!-- Hide cell code -->
  <script src="/the-fair-cookbook/assets/js/page/hide-cell.js" async></script>

  <!-- Printing the screen -->
  <!-- Include nbinteract for interactive widgets -->
<script src="https://printjs-4de6.kxcdn.com/print.min.js" async></script>
<script>
printContent = () => {
    // MathJax displays a second version of any math for assistive devices etc.
    // This prevents double-rendering in the PDF output.
    var ignoreAssistList = [];
    assistives = document.querySelectorAll('.MathJax_Display span.MJX_Assistive_MathML').forEach((element, index) => {
        var thisId = 'MathJax-assistive-' + index.toString();
        element.setAttribute('id', thisId);
        ignoreAssistList.push(thisId)
    });

    // Print the actual content object
    printJS({
        printable: 'textbook_content',
        type: 'html',
        css: "/the-fair-cookbook/assets/css/styles.css",
        style: "#textbook_content {padding-top: 40px};",
        scanStyles: false,
        targetStyles: ["*"],
        ignoreElements: ignoreAssistList,
        documentTitle: "Made with Jupyter Book"
    })
};

initPrint = () => {
    document.querySelector('#interact-button-print').addEventListener('click', printContent)
}

initFunction(initPrint)
</script>

</head>

  <body>
    <!-- Include the ThebeLab config so it gets reloaded on each page -->
    <script type="text/x-thebe-config">{
    requestKernel: true,
    binderOptions: {
    repo: "nih-cfde/the-fair-cookbook",
    ref: "gh-pages",
    },
    codeMirrorConfig: {
    theme: "abcdef",
    mode: "python"
    },
    kernelOptions: {
    kernelName: "python3",
    path: ""
    }
}
</script>

    <!-- .js-show-sidebar shows sidebar by default -->
    <div id="js-textbook" class="c-textbook js-show-sidebar">
      



<nav id="js-sidebar" class="c-textbook__sidebar">
  <a href="https://nih-cfde.github.io/the-fair-cookbook/intro.html"><img src="/the-fair-cookbook/images/logo/CFDE-logo-contrast.png" class="textbook_logo" id="sidebar-logo" alt="textbook logo" data-turbolinks-permanent/></a>
  <h2 class="c-sidebar__title">NIH-CFDE FAIR COOKBOOK</h2>
  <ul class="c-sidebar__chapters">
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/intro">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/intro.html"
        >
          
          Introduction
        </a>
      </li>

      
      

      

      
      

      
        

        

        <ul class="c-sidebar__sections u-hidden-visually">
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/00/fair-principles">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/00/fair-principles.html"
              >
                
                What is FAIR?
              </a>
            </li>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/05/cfde">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/05/cfde.html"
              >
                
                Engaging with NIH-CFDE
              </a>
            </li>
            
            
          
        </ul>
      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/recipes/recipes-overview">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/recipes/recipes-overview.html"
        >
          
          Recipes
        </a>
      </li>

      
      

      

      
      

      
        

        

        <ul class="c-sidebar__sections u-hidden-visually">
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/06/cataloguing">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/06/cataloguing.html"
              >
                
                Discoverability
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/12/c2m2">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/12/c2m2.html"
                  >
                    
                    Introducing the Cross-Cut Metadata Model (C2M2)
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/12/C2M2-L1-description">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/12/C2M2-L1-description.html"
                  >
                    
                    Description of the Level-1 C2M2
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/06/cfde-namespaces">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/06/cfde-namespaces.html"
                  >
                    
                    CFDE Resource Naming
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/07/seo">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/07/seo.html"
                  >
                    
                    ETL to C2M2/DATS - the KidsFirst example
                  </a>
                </li>
              
              </ul>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/08/identifiers">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/08/identifiers.html"
              >
                
                Identification
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/08/1/pids">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/08/1/pids.html"
                  >
                    
                    Minting persistent and resolvable identifiers
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/08/2/minids">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/08/2/minids.html"
                  >
                    
                    Using MINIDs
                  </a>
                </li>
              
              </ul>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/09/finding">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/09/finding.html"
              >
                
                Semantics
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/10/ontologies">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/10/ontologies.html"
                  >
                    
                    Understanding terminologies
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/11/onto-services">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/11/onto-services.html"
                  >
                    
                    Ontology services
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/14/cfde-terminologies">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/14/cfde-terminologies.html"
                  >
                    
                    NIH CFDE selected terminologies
                  </a>
                </li>
              
              </ul>
            
            
          
            
            

            
            
            
            

            <li class="c-sidebar__section" data-url="/recipes/fair-assessment">
              <a class="c-sidebar__entry"
                href="/the-fair-cookbook/recipes/fair-assessment.html"
              >
                
                FAIR compliance
              </a>
            </li>
            
              
              <ul class='c-sidebar__subsections u-hidden-visually'>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/04/fairshake">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/04/fairshake.html"
                  >
                    
                    Evaluating FAIRness with FAIRshake
                  </a>
                </li>
              
                
                
                
                
                <li class="c-sidebar__subsection" data-url="/recipes/16/fair-api">
                  <a class="c-sidebar__entry"
                    href="/the-fair-cookbook/recipes/16/fair-api.html"
                  >
                    
                    Developing FAIR API
                  </a>
                </li>
              
              </ul>
            
            
          
        </ul>
      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/CFDE_Glossary">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/CFDE_Glossary.html"
        >
          
          Glossary
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/contributing">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/contributing.html"
        >
          
          Contributing
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/governance">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/governance.html"
        >
          
          Governance
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/license">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/license.html"
        >
          
          License
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/disclaimer">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/disclaimer.html"
        >
          
          Disclaimer
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="/code_of_conduct">
        <a class="c-sidebar__entry"
          href="/the-fair-cookbook/code_of_conduct.html"
        >
          
          Code of Conduct
        </a>
      </li>

      
      

      

      
      

      

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      

      
      
      
      

      
      
      <li class="c-sidebar__chapter" data-url="https://github.com/nih-cfde/the-fair-cookbook">
        <a class="c-sidebar__entry"
          href="https://github.com/nih-cfde/the-fair-cookbook"
        >
          
          GitHub repository
        </a>
      </li>

      
      

      

      
      

      

      
    
  </ul>
  <p class="sidebar_footer">Powered by <a href="https://jupyterbook.org">Jupyter Book</a></p>
</nav>

      
      <div class="c-topbar" id="top-navbar">
  <!-- We show the sidebar by default so we use .is-active -->
  <div class="c-topbar__buttons">
    <button
      id="js-sidebar-toggle"
      class="hamburger hamburger--arrowalt is-active"
    >
      <span class="hamburger-box">
        <span class="hamburger-inner"></span>
      </span>
    </button>
    <div class="buttons">
<div class="download-buttons-dropdown">
    <button id="dropdown-button-trigger" class="interact-button"><img src="/the-fair-cookbook/assets/images/download-solid.svg" alt="Download" /></button>
    <div class="download-buttons">
        
        <a id="interact-button-print"><button id="interact-button-download" class="interact-button">.pdf</button></a>
    </div>
</div>


</div>

  </div>
  <!-- Empty sidebar placeholder that we'll auto-fill with javascript -->
  <aside class="sidebar__right">
    <header><h4 class="nav__title"><img src="/the-fair-cookbook/assets/images/list-solid.svg" alt="Search" />   On this page</h4></header>
    <nav class="onthispage">
    </nav>
  </aside>
  <a href="/the-fair-cookbook/search.html" class="topbar-right-button" id="search-button">
    <img src="/the-fair-cookbook/assets/images/search-solid.svg" alt="Search" />
  </a>
</div>

      <main class="c-textbook__page" tabindex="-1">
            <div class="c-textbook__content" id="textbook_content">
              <h1 id="cfde-glossary">CFDE Glossary</h1>

<h2 id="controlled-vocabulary">Controlled Vocabulary</h2>

<p><strong>Asset</strong> - a sample or a file.</p>

<p><strong>Biospecimens</strong> - a material collected from an organism, a cell culture, or a material containing organisms, such as an environmental material.</p>

<p><strong>Data-Generating Events</strong> - a process, such as a data acquisition or a data transformation, resulting in the creation of a file (digital asset).</p>

<p><strong>Data Object</strong> - a region of storage that contains a value or group of values. Each value can be accessed using its identifier or a more complex expression that refers to the object. In addition, each object has a unique data type.</p>

<p><strong>Dataset</strong> - a collection of data, published or curated by a single agent, and available for access or download in one or more formats.</p>

<p><strong>DATS</strong> - <a href="https://github.com/datatagsuite/README">Data Article Tag Suite</a> is a data model for representing key information about datasets with an emphasis on data discovery and data findability, which has inspired the creation of the NIH-C2M2 model. <a href="https://datatagsuite.github.io/docs/html/dats.html">The DATS model is expressed as a JSON schema</a>. Associated JSON-LD context files support search engine optimization because they map into <a href="https://schema.org">schema.org</a> and <a href="www.w3.org/TR/vocab-dcat-2/">DCAT</a>. Mappings into biological entities are also available via <a href="http://www.obofoundry.org">OBO Foundry resources</a>.</p>

<p><strong>DERIVA</strong> - Discovery Environment for Relational Information and Versioned Assets. A suite of tools and services that are designed to significantly reduce the overhead and complexity of creating and managing complex, big datasets. DERIVA provides a digital asset management system for scientific data to streamline the acquisition, modeling, management, and sharing of complex, big data, and provides interfaces so that these data can be delivered to diverse external tools for big-data analysis and analytic tools.</p>

<p><strong><a href="https://www.go-fair.org/fair-principles/">FAIR</a></strong> - Findable, Accessible, Interoperable, and Reusable. Making data FAIR is the main goal of the DCCs.</p>

<p><strong><a href="https://fairshake.cloud/">FAIRshake</a></strong> - a tool produced for carrying out <em>FAIR Assessment</em>. Under the CFDE, each data center’s inventory will be evaluated consistently based on FAIRshake, and the Coordinating Center will work with the individual Common Fund programs to adjust FAIR measures to meet the needs of the Common Fund. FAIRshake includes evaluating the FAIRness of digital objects including datasets, tools, and repositories.</p>

<p><strong>Figshare</strong> - an online open access repository where researchers can preserve and share their research outputs, including figures, datasets, images, and videos.</p>

<p><strong>File</strong> - digital objects that each of the DCCs host, such as genomic sequence, metagenomic, RNA-Seq, physiological, and metabolic data.</p>

<p><strong>Frictionless Data Package</strong> - a format specification produced by the Open Knowledge Foundation and supported by the Frictionless.io organization. It aims to shorten the path from data to insight with a collection of specifications and software for the publication, transport, and consumption of data. This kills the cycle of find/improve/share that makes for a dynamic and productive data ecosystem.</p>

<p><strong>Globus Automate</strong> - a distributed research automation platform that addresses the problem of securely and reliably automating, for many thousands of scientists, sequences of data management tasks that may span locations, storage systems, administrative domains, and timescales, and integrate both mechanical and human inputs.</p>

<p><strong>GUID</strong> - Globally Unique IDentifier. A 128-bit unique identifier used to uniquely identify an entity as modern hashing functions used to generate those make identifier collisions (the event of the function producing the same sequence) highly unlikely (but not impossible).</p>

<p><strong>Insignia</strong> - a grid of colored squares used to visually communicate FAIRness level.</p>

<p><strong>Metadata</strong> - a type of information entity usually defined as data about the data, understood as descriptors to understand the context of a dataset. For example, metadata about an FASTQ file may be file size or file creator. Metadata is often classified into descriptive metadata, structural metadata, administrative metadata, and provenance metadata, all of which provide context to the actual data/dataset.</p>

<p><strong>Organization</strong> - an entity comprising of multiple people, such as an institution or an association, that has a particular purpose.</p>

<p><strong>POC I/II</strong> - proof of concept. A realization of a certain method or idea in order to demonstrate its feasibility.</p>

<p><strong>Project</strong></p>

<p><strong>Sample</strong> - see <em>Biospecimens</em>.</p>

<p><strong>Subject</strong> - a study participant (human, animal) from which samples may be obtained.</p>

<p><strong>Subject Group</strong> - a set of study subjects, often grouped based on a set of criteria or the types of intervention study that the subjects will undergo.</p>

<h2 id="file-formats-and-types">File Formats and Types</h2>

<p><strong>CFDE Asset Manifest</strong> - a collection of <em>Assets</em> described by the <em>CFDE Asset Specification</em>. The ecosystem will support the concept of a manifest that describes a collection of files. The manifests enable bundling lists of CFDE data assets into a machine-readable file using a common format. Manifests will also be used to publish the complete inventories of data from each DCC, and will enable uniform collection of asset metadata to support indexing of the assets in the CFDE portal.</p>

<p><strong>CFDE Asset Specification</strong> - defines the set of attributes used to charaterize an <em>Asset</em>. The specification simplifies the discovery of assets hosted at the DCCs with a minimal set of descriptors for each of these files. The types of files that are referenced (e.g., genomic sequence, metagenomic, RNA-Seq, physiological and metabolic data) are flexible and contain a small number of essential elements such as a <em>GUID</em>, originating institution (e.g., Broad Institute), assay type (e.g., whole genome/exome, transcriptome, epigenome), file type (e.g., fastq, alignment, vcf, counts), and tissue source and species name for the sample.
<img src="https://user-images.githubusercontent.com/40363469/66134046-ac16bc80-e5c5-11e9-9b30-66407a3446e5.png" alt="data asset specification" /></p>

<p><strong><a href="https://fair-research.org/deliverables/cfde-metadata-format.html">CFDE Common Metadata Format</a></strong> - specifies a minimal set of attributes (metadata) related to an <em>Asset</em>.
<img src="https://user-images.githubusercontent.com/40363469/66134031-a8833580-e5c5-11e9-9cc8-5e4b275fa20c.png" alt="common metadata format" /></p>

<p><strong><a href="https://github.com/nih-cfde/cfde-deriva/blob/2019-08/diagrams/cfde-core-model.png">Core ER Diagram</a></strong></p>

<p><strong><a href="https://github.com/nih-cfde/cfde-deriva/blob/2019-08/table-schema/cfde-core-model.json">Core Table Schema</a></strong></p>

<p><strong><a href="https://github.com/nih-cfde/cfde-deriva/blob/2019-10/extractors_and_metadata.GTEx/cfde-core-model.2019.10.21.1430.png">Entity-Relation Diagram</a></strong></p>

<p><strong>Inventory</strong> - asset inventory distributed by the DCCs through a portal.</p>

<p><strong>Metadata Manifest</strong> - a file that includes metadata for a group of accompanying files that are part of a coherent unit (manifest), such as name, version, background scripts, and browser actions.</p>

<h2 id="processes">Processes</h2>

<p><strong>C2M2 Creation Flow</strong></p>

<p><strong>C2M2 Extractor Flow</strong></p>

<p><strong>FAIR Assessment</strong> - the process of evaluating the level of compliance of a dataset with the FAIR principles. The assessment may be performed with tools such as <a href="https://fairshake.cloud">FAIRshake</a> or <a href="https://fairsharing.github.io/FAIR-Evaluator-FrontEnd/#!/">FAIREvaluator</a>.</p>

<p><strong>Globus Automate Flow</strong> - Globus, a software service created and based at the University of Chicago, helps scientists simplify their workflow by automating data transfer and synchronization tasks. Users can create automated sequences initiated by events.</p>

<p><strong>Master Flow</strong></p>

<p><strong>Metadata Ingest</strong> - assigning identifiers to the objects and then extracting or creating metadata for these objects.</p>

<p><strong><a href="https://github.com/nih-cfde/cfde-deriva/blob/2019-08/examples/tableschema_to_deriva.py">Table Schema to DERIVA Translation</a></strong></p>

<h2 id="software-and-tools">Software and Tools</h2>

<p><strong>API</strong> - Application Programmer Interface. Allows developers to manipulate (query, update) remote data sources through specific protocols or specific standards for communication (e.g., REST,SOAP). An important element of the ecosystem will be the standardization and publishing of an API that can be used by data consumers to retrieve the inventories, the data asset specification, and additional metadata associated with the assets. This will allow for consumers of these inventories to programmatically interrogate the federated system for information that may be relevant to a consuming service.</p>

<p><strong>C2M2</strong> - Cross Cut Metadata Model.</p>

<p><strong><a href="https://cfde.derivacloud.org/deriva-webapps/plot/">CFDE Data Dashboard</a></strong> - an interface that monitors DCC data upload to the cloud and usage statistics to support cross-DCC search and ecosystem integration.</p>

<p><strong>CFDE Query Portal</strong> - a portal enabling users and administrators to search all the federated data assets at each Common Fund Program. The CFDE portal increases a user’s ability to find these important resources, as well as mix and match sets of data from each site to use in subsequent analysis. Administrators and Program Officers at Common Fund can use a single website to view the growth of data from their program over time, review objective FAIR metrics for these assets, understand download statistics and geographic distribution, and view the degree of harmonization of these data in comparison to other sites.</p>

<p><strong><a href="https://github.com/nih-cfde/FAIR/blob/master/Demos/FAIRAssessment/data_citation_rubric.py">Data Citation Rubric</a></strong></p>

<p><strong>dbGaP</strong> - Database of Genotypes and Phenotypes.</p>

<p><strong>DERIVA Catalog</strong> - a group of assets that may be digital objects (i.e., files) or references to physical objects. DERIVA uses an entity-relationship data model to catalog and organize these assets.</p>

<p><strong>FAIR Insignia</strong> - to visually communicate FAIRness level, a grid of colored squares, called the FAIR insignia, was developed. The FAIR insignia identifies areas of strength and weakness in the FAIRness level of digital objects, guiding digital object producers on how to improve the FAIRness of their products.</p>

<p><strong>GitHub</strong> - an online hub for storing and sharing computer programs and other plain text files. The CFDE team uses it for storage, hosting websites, communications, and project management.</p>

<p><strong>Jupyter Notebook</strong> - a web-based interactive environment for organizing data, performing computation, and visualizing output.</p>

<p><strong>NPM</strong> - a package manager for the JavaScript programming language. It is the default package manager for the JavaScript runtime environment Node.js. It consists of a command line client, also called npm, and an online database of public and paid-for private packages, called the npm registry.</p>

<h2 id="use-case-terms">Use Case Terms</h2>

<p><strong>Objective</strong> - a description of a scientific process, told without reference to a specific technology solution. Focuses on resources required, method(s) followed, and outcome(s) targeted. Can be validated with scientific stakeholders.</p>

<p><strong>Persona</strong> - a type of user who will appear in the epics and stories that follow.</p>

<p><strong>Requirement</strong> - a well-scoped and defined piece of software or data engineering work that is needed to support a User Task. These should be small tasks that can be verified with engineering teams (e.g., did we deliver it? Yes or no.) A single requirement may be important for any number of different User Tasks, and various collections of Requirements can be grouped to support different User Tasks.</p>

<p><strong>Summary</strong> - a high-level, non-technical description of an entire Use Case. The user in each summary has a name, a scientific or administrative problem, and both proximate and ultimate goals. The focus is on the problem and what is generally needed to solve it.</p>

<p><strong>User Task</strong> - a story, told from the user’s perspective that captures a concrete step in a user’s interactions with tools (e.g., software solutions) in the service of achieving the scientific objective. Must be written in terms that are meaningful to the user, from their perspective. This can be thought of as one in a series of medium scale tasks that must be completed to answer the question posed in the scientific objective. The list of User Tasks in a Use Case should cover everything the users needs to achieve their goals, even interactions that do not involve the software or data from the Common Fund.</p>

<h2 id="other">Other</h2>

<p><strong>A2CPS</strong> - NIH’s Acute to Chronic Pain Signatures program. The most comprehensive study to date to investigate the connections of peripheral biology, brain, psychological, and bio-behavioral risk factors. Composed of a consortium of organizations and scientists throughout the U.S. and Canada, A2CPS is part of the multi-pronged <a href="https://heal.nih.gov/">NIH HEAL</a> (Helping to End Addiction Long-Term) initiative, an aggressive effort to speed scientific solutions to stem the national opioid public health crisis.</p>

<p><strong>ASHG</strong> - American Society of Human Genetics. They hold an annual conference that is the largest human genetics and genomics meeting and exposition in the world.</p>

<p><strong>BIC</strong> - Bioinformatics Center. The BIC is responsible for the task of harmonizing the large quantity and diversity of data and metadata being generated by the consortium and performing meaningful integrative analyses across these omics data types.</p>

<p><strong>CC</strong> - Coordinating Center. The CFDE-CC’s purpose is to improve growth in use and reuse of CF program data by supporting DCCs that participate in the Common Fund Data Ecosystem.</p>

<p><strong>CFDE</strong> - Common Fund Data Ecosystem. A data ecosystem is a collection of data silos or commons joined together by a set of standards and services that facilitate findability, accessibility, reuse, and interoperability of datasets between silos/commons. A data ecosystem is focused on enabling multi-way connectivity between datasets, in a horizontal fashion, rather than deeper vertical analysis within each dataset. The goal of an ecosystem is to enable use cases between data silos, not within.</p>

<p><strong>Common Fund Programs</strong> - the intent of the CF programs is to provide a strategic and nimble approach to address key roadblocks in biomedical research that impede basic scientific discovery and its translation into improved human health. In addition, these programs capitalize on emerging opportunities to catalyze the rate of progress across multiple biomedical fields. The CF programs include:</p>
<ul>
  <li>4DN - The 4D Nucleome program’s goal is to study the three-dimensional organization of the nucleus in space and time (the 4th dimension).</li>
  <li>GTEx - the Genotype-Tissue Expression project is an ongoing effort to build a comprehensive public resource to study tissue-specific gene expression and regulation.</li>
  <li>HMP - the Human Microbiome Project whose mission is to generate resources to facilitate characterization of the human microbiota to further our understanding of how the microbiome impacts human health and disease.</li>
  <li>Kids First - the Kids First Pediatric Research Program Data Resource Center enables researchers, clinicians, and patients to work together to accelerate research and promote new discoveries for children affected with cancer and structural birth defects.</li>
  <li>LINCS - the Library of Integrated Network-based Cellular Signatures project is based on the premise that disrupting any one of the many steps of a given biological process will cause related changes in the molecular and cellular characteristics, behavior, and/or function of the cell – the observable composite of which is known as the cellular phenotype. Observing how and when a cell’s phenotype is altered by specific stressors can provide clues about the underlying mechanisms involved in perturbation and, ultimately, disease.</li>
  <li>Metabolomics - this program was developed with the goal of increasing national capacity in metabolomics by supporting the development of next generation technologies, providing training and mentoring opportunities, increasing the inventory and availability of high quality reference standards, and promoting data sharing and collaboration.</li>
  <li>SPARC - the Stimulating Peripheral Activity to Relieve Conditions program seeks to accelerate development of therapeutic devices that modulate electrical activity in nerves to improve organ function.</li>
</ul>

<p><strong>Controlled Vocabulary</strong> - an organized arrangement of words and phrases used to index content and/or to retrieve content through browsing or searching. It typically includes preferred and variant terms and has a defined scope or describes a specific domain. For example, the DCIC curates an internal 4DN controlled vocabulary to provide definitions for emerging technologies and techniques, metadata terms, and captures important data features not defined by previous ontologies.</p>

<p><strong>DCC</strong> - Data Coordinating/Resource Center.</p>

<p><strong>DCIC</strong> - Data generated by 4DN partner institutions are integrated, curated, analyzed, and disseminated by the 4DN Data Coordination and Integration Center (DCIC).</p>

<p><strong>DTC</strong> - Data and Tools Cores. The Metabolomics Program consortium consists of six RCMRCs and seven DTCs that are overseen by the Metabolomics Consortium Coordinating Center at the University of Florida.</p>

<p><strong>eRA Commons</strong> - Designated ID provider for the whitelist.</p>

<p><strong>HMP</strong> - Human Microbiome Project.</p>

<p><strong>M3C</strong> - the Metabolomics Consortium Coordinating Center at the University of Florida, which handles overall coordination for the Metabolomics program.</p>

<p><strong>Metabolomics Workbench Metabolite Database</strong> – a Postgres database containing over 65,000 structures and annotations of biologically relevant metabolites collected from public repositories (e.g., LIPID MAPS, ChEBI, HMDB, BMRB, PubChem, KEGG). Users can search for metabolites in the database by substructure, text, or mass (m/z ratio). Each entry contains key information about the metabolite, including structure, molecular weight, common and systematic names, PubChem compound ID, and classification. Entries also contain cross references to external databases and repositories (e.g., HMDB, ChEBI, LIPID MAPS, METLIN, ChemSpider, KEGG, etc.) as well as links to the MoNA MS spectra and human metabolic pathways containing the metabolite. Additionally, the open-source chemistry cartridge enables substructure searching, generation of chemistry-centric attributes (formula, exact mass), and interconversion of molecular formats.</p>

<p><strong>MGP</strong> – Human Metabolome Gene/Protein Database. Developed by the Common Fund Metabolomics program. A database of metabolome-related genes and proteins containing over 7,300 genes and over 15,500 proteins. Users can search by gene (name, symbol, entrez ID, etc.), HMDB Pathway, or Reactome Pathway. MGP displays genes/proteins and metabolites associated with a pathway of interest. Searching by gene displays information about the gene’s associated proteins and pathways, including a summary of the function and metabolites involved in the pathway.</p>

<p><strong>MoTrPAC</strong> – Molecular Transducers of Physical Activity Consortium.</p>

<p><strong>MW</strong> – Metabolomics Workbench. Developed by the Common Fund Metabolomics program, the MW is an online interface to the NMDR developed at UCSD. It allows users to manage and upload studies as well as browse and search available studies. Using the MW interface, submitters upload data and results, including metadata, targeted data measurements, protocols/methods files, untargeted data measurements, and raw data (MS/NMR files, etc.). Other researchers can then use the MW website to browse, search, analyze, and download data as well as view summary figures of key study search parameters (e.g., bubble chart showing studies by sample source). For example, studies can be filtered by study metadata (disease, sample source, species, instrumentation) or metabolite information (metabolite classification, biochemical pathways, retention time, etc.) to identify data relevant to the user’s needs. Additionally, it provides analysis tools and access to metabolite standards, protocols, tutorials, training, and other resources to support metabolomic researchers.</p>

<p><strong>NMDR</strong> – National Metabolomics Data Repository. Responsible for collating, analyzing, and distributing the data gathered by the RCMRCs and hosting the tools and methods created by the DTCs.</p>

<p><strong>RAS</strong> – Researcher Auth Service. A service under development by the NIH’s Center for Information Technology that will facilitate access to controlled data assets and repositories.</p>

<p><strong>RCMRC</strong> – Regional Comprehensive Metabolomics Resource Cores. The Metabolomics Program consortium consists of six RCMRCs, also called the Compound Identification Cores (CIDCs), and seven Data and Tools Cores (DTCs) that are overseen by the Metabolomics Consortium Coordinating Center at the University of Florida.</p>

<p><strong>RefMet</strong> – Reference list of Metabolite names. Developed by the Common Fund Metabolomics program. Effectively a large spreadsheet that provides a standard nomenclature for over 95,500 chemical species. From the Metabolomics Workbench website, it can be browsed and searched directly or a user can input a list of metabolite names and have them automatically converted to RefMet nomenclature. A user can also directly download the data, either in whole or after filtering as one would with a simple Excel sheet. Or the entire dataset can be downloaded as part of a Shiny R app and queried locally.</p>

<p><strong>SSO</strong> – Single Sign-On.</p>

<p><strong>STRIDES</strong> – NIH’s Science and Technology Research Infrastructure for Discovery, Experimentation, and Sustainability initiative. Common Fund leadership has partnered with the STRIDES initiative, which provides lower-cost cloud services to NIH projects.</p>

<p><strong>TCC</strong> – Training Coordination Center. This center is staffed by experts in bioinformatics curriculum development, teaching, and community building. It provides support and resources for the development of DCC-specific training programs as well as end-user training on CFDE products and general topics of interest to the Common Fund research community. The TCC can help with logistical support for hosting workshops, as well as providing guidance on how to grow and build a sustainable training program. The TCC provides instructor training for the DCCs and assists with creating useful qualitative and quantitative feedback and assessment tools. In addition to site-specific training, the TCC offers training on CFDE products as they become available, and pilots a general bioinformatics workshop curriculum on topics of broad interest within the Common Fund.</p>

<p><strong>TOPMed</strong> – Trans-Omics for Precision Medicine.</p>

<h1 id="draft---list-of-term-anchors-currently-linked-from-spec-documents">DRAFT - list of term anchors currently linked from spec documents</h1>

<p><strong>CFDE-asset-manifest</strong></p>

<p><strong>DCCs</strong></p>

<p><strong>DCC-data-ingestion-process</strong></p>

<p><strong>Entity-relationship-model</strong></p>

<p><strong>Metadata</strong></p>

<p><strong>Richness-levels</strong></p>

            </div>
            <div class="c-textbook__footer" id="textbook_footer">
              
<nav class="c-page__nav">
  

  
</nav>

              <footer>
  <p class="footer">This project is supported by NIH CFDE program <a href="https://commonfund.nih.gov/dataecosystem"><img src="/the-fair-cookbook/assets/images/logo/CFDE-logo.png" height="20" alt="NIH flag" class="float-left bottom-margin"></a> <script>mermaid.init();</script></p>
</footer>

            </div>

        </div>
      </main>
    </div>
  </body>
</html>
